---
title: "Homework 7: Hypergeometric test and line plots"
author: "Ozan Berk Imir"
date: "Due 10/22/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## 1) Gene set overlap

A recent study, [***Genetically engineered cerebral organoids model brain tumor formation*** (Bian et al., *Nat Meth* 2018)](https://www.nature.com/articles/s41592-018-0070-7),
created organoid models of two brain tumor types (primitive neuroectodermal
tumors and glioblastoma) using CRISPR to generate mutations in known oncogenes. 
The two conditions were generated by MYC oncogene overexpression, *MYC^OE^* (CNS-NET), and knockouts of several genes including *PTEN* and *P53* (GBM).

Differences in gene expression between two tumor models were compared with controls
and were found to differ in the sets of genes that were mis-regulated. 
The researchers first showed that the two organoid tumor models seem to show quite different
gene expression patterns using **PCA** (principal components analysis) of the top 500
genes with the highest variability compared with controls (see [**Figure 3a,b**](https://www.nature.com/articles/s41592-018-0070-7/figures/3)). 
We will go over PCA later in the course.

Next, they performed **hypergeometric tests** to ask whether all of the genes that 
were significantly **up- or down-regulated** in the two models (in comparison with 
controls) showed less overlap that expected by random chance (see [**Supplementary Figure 3**](https://www.nature.com/articles/s41592-018-0070-7/figures/9)).

If so, this would support the idea that divergent gene programs are misregulated 
in the two models.

Based on the PCA analysis (**Figure 3a**), differentially expressed genes could be separated into two different groups corresponding to the two organoid tumor models: Cluster 2 (*MYC^OE^*) and Cluster 3 (GBM).

![](Bian_Fig3a.png){width=50%}

**Fig. S3** shows Venn diagrams of the gene set overlaps between Clusters 2 and 3:

![](Bian_FigS3a.png){width=50%}

Your question is whether the differentially expressed genes in these two clusters overlap by **less** than expected by chance. Here we will focus on just the **up-regulated genes**. How will you set up the problem?


### a) Expected gene set overlap based on independence

In total, 4034 genes were up-regulated in at least one of the organoid tumor models (vs. controls). The rest of the numbers you will need are shown in the Venn diagrams in Fig. S3. A framework for organizing the data has been set up below to help get you started.

```{r, collapse=T}
## Up-regulated genes ======================================================== #
## 4034 genes up-regulated in either experiment (vs control)

ol.obs =  92    # overlap
A =  2155 + 92  # total size of Cluster 2 (Set A = bigger set)
B =  203 + 92   # total size of Cluster 3 (Set B = smaller set)
N =  4034       # total genes up-regulated in both experiments
# 
# # What is the expected overlap based on the null hypothesis of independence?
ol.exp = (A * B)/N
ol.exp
```


### b) Hypergeometric test

Find a p-value for the observed overlap with a hypergeometric test using `dhyper` and `pyhper`. Remember, you are looking for less overlap than expected between the two sets (is this the lower tail or the upper tail?).

First, look up the documentation on these functions.
```{r}
?dhyper
?phyper
```

Here are some additional tips to help you figure out how to parameterize the functions:

+ $x$: this is the observed overlap
+ $m$: pick one gene set, call this Set A
+ $n$: this is the total number of genes NOT in Set A (out of all the up-regulated genes)
+ $k$: the other set (Set B)
  +  Note that the maximum possible size of the overlap would be the total size of the smaller gene set (which cluster is that here?)
  + It does not really matter which set you choose to use for $m$ or $k$; the p-value will be the same either way.

```{r, collapse=T}
## p-value using hypergeometric test ========================================= #

# using CDF function
phyper(ol.obs, B, N - B, A)
# using PDF function
sum(dhyper(1:ol.obs, B, N - B, A))
```


### c) Fisher's Exact Test

The hypergeometric test is equivalent to a one-tailed Fisher's test (but it is 
actually more efficient in R). Use a Fisher's test to make the same comparison as above. Here are some helpful hints:

+ You want to perform a one-sided test for more extreme values than the ones observed, given all possible combinations of values "conditioned on fixed margins" (which tail are you looking for here?).
+ Make a contingency table: orientation is arbitrary, but by convention put the categories "of interest" first.
  + Put the overlap in the top left cell (*a*, a.k.a. $x_{11}$).
  + Put the remainder of the A and B sets (minus the overlap) in the right (*b*, $x_{12}$) and bottom left (*c*, $x_{21}$) cells.
  + Put the remainder of the up-regulated genes in (*d*, $x_{22}$).

```{r, collapse=T}
## with Fisher's exact test ================================================== #
A.not.B = A-ol.obs
B.not.A = B-ol.obs
N.not.AB = N - B - A + ol.obs  # same as: N - B - A.not.B
contingency.table = rbind(c(ol.obs, A.not.B),
                          c(B.not.A, N.not.AB))
rownames(contingency.table) = c("down.overlap","down.not-overlap")
colnames(contingency.table) = c("up.overlap","up.not-overlap")
contingency.table
fisher.test(contingency.table, alternative="less")
```

### d) How do the p-values compare?

```{r}
# The p-value for the Fisher's Exact Test is larger than the p-value calculated for the hypergeometric tests, but as the Fisher's Exact Test is a bit more accurate when deducing all the different contingencies that are greater or lesser than the ones observed, the difference in p-values is to be expected. 
```


## 2) Line plots: Mean, SD, SEM

Please read the short article *Error bars (Nat Meth 2013)* provided with this assignment.

### a) What is the difference between the standard error (SEM) of the mean and standard deviation?  
```{r eval=FALSE}
# Error bars based on standard deviation (SD) talk about the spread of the population whereas the error bars based on the SEM talk about the uncertainty of the population. If one uses SD as the error bars, then the variance is indicated, not the error. Consequently, for smaller samples, non-overlapping SEM error bars do not mean the difference between these two groups is significant. 
```

### b) What are some common misconceptions about interpreting error bars?
```{r eval=FALSE}
# 1) Non-overlapping SEM error bars mean significant difference (FALSE for small sample size)
# 2) CI captures the mean when a second sample is taken from the same population with a percent chance of the CI (FALSE as all sample means vary so the chance is lower)
```

## 3) Cancer metabolism

In the paper [***Metabolic competition in the tumor microenvironment is a driver of cancer progression*** (Chang et al., *Cell* 2015)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4864363/), the authors use mouse models to demonstrate that in cancers, T cells are metabolically restricted by tumor glucose consumption, ultimately allowing tumor progression. The authors use line graphs in **Figure 4B**:

![](Chang_Fig4B.png){width=60%}

### a) What question were the researchers trying to answer in this Figure?
```{r eval=FALSE}
# They were interested in how the tumor sizes would change over time after the transplantation of the tumor cells. So the question would be "how fast do tumors grow with different blockade antibody treatments?"
```

### b) How many replicates were performed, and what do the error bars represent?
```{r eval=FALSE}
# 6-8 animals were included for each treatment group and the experiments were repeated 1-3 times. The error bars represent the SEM for 3+ experiments. 
```

### c) Do you think this an appropriate way to represent the data? Why or why not, and how would you try to represent them?
```{r eval=FALSE}
# In animal experimentation, the number of animals selected generally is chosen based on how large the animal is, how closely it represents the human genome and physiology and how costly it is to house the animals. In this case, as mice are good representatives when it comes to tumor progression, 6-8 mice per group and 3+ experimental repeats which help form a significance. As the results for days 15+ are very far apart, I think using SEM as error bars makes sense even though it is usually misleading for small sample sizes. 
```
    

## 4) Spindle dynamics in evolution

The paper [***Stoichiometric interactions explain spindle dynamics and scaling across 100 million years of nematode evolution*** (Farhadifar et al., *eLife* 2020)](https://elifesciences.org/articles/55877) explores spindle dynamics in *C. elegans* using a combination of imaging methods and QTL mapping of intercross lines. Below is reproduced Figure 3 from this article:

![](Farhadifar_Fig3.jpg){width=80%}

### a) What question were the researchers trying to answer in this figure?
```{r eval=FALSE}
# They wanted to map the quantitative trait loci, meaning the change in certain genetic traits mapped to a certain trait, in order to understand which genomic locations, when changed, would alter certain traits and to what degree this change would occur. The question for this particular figure is: "which markers change final spindle length and how does the inhibition of these regulators affect spindle length?" 
```

### b) Panels C and D

About how many data points are there for each line? What do the error bars represent? Would you change anything about the way the data are represented?
```{r eval=FALSE}
# 500 permutations were performed for each line. The error bars represent the standard error centered around the mean of each allele (N2 and CB4856). The error bars are small enough and there are ample data points to retain the measure of standard error of the mean in order to represent a difference between the alleles given different QTLs (QTL1 vs QTL2)
```

### c) Panels E and F

What is being plotted? What do the shaded areas represent, and why do you think the researchers chose to display the variation in their measurements in this way? Would you have done anything differently?
```{r eval=FALSE}
# The figure quantifies the spindle length as a function over time when gpr-1, a gene involved in astral microtubule elongation in embryos; and par-2, a gene in the PAR system involved in cytoplasmic partitioning, are inhibited through RNAi. gpr-1 matches up with QTL1 and par-2 matches up with QTL2 in location, so their involvement in spindle length elongation is hypothesized to be significant. The shaded areas represent the error bars, so instead of having data points with error bars, they represented the error range as a shaded area, because they use live imaging and have many data points over time. I think this representation makes sense when looking at the end of the experiment, but when the difference between the two groups (control and RNAi) are significant is unclear, so I would prefer to have discrete data points over time with a set time interval to match this data such that we can more clearly pinpoint that instance. 
```


## 5) Line plots with SD or SEM

How would you plot similar graphs in R showing SD or SEM? The dataset "HW7_data.csv" contains eight replicate measurements across 20 timepoints (0-190 seconds) for a control and a treatment condition. The data are the raw (non-normalized) values used to generate the plots in Figure 6A from 
[Cipriani et al., eLife 2021](https://elifesciences.org/articles/60833).

### a) Calculate the mean, SD, SEM

Load the dataset and use the `dplyr` package to reformat it as follows:

+ Group the data by treatment and time (hint: use `group_by()`).
+ Add 3 new columns to the data frame containing group summaries for the mean, SEM, and SD (hint: use `summarise()`).
+ Add 4 new columns containing the upper and lower limits for the SEM and the SD.
  
```{r}
library(dplyr)
c.elegans = read.csv(file = "HW7_data.csv", header = TRUE)
c.elegans

c.elegans_dplyr = c.elegans %>%
  group_by(TREATMENT, time) %>%
  summarize(group_mean = mean(Values),
            group_SEM  = sd(Values, na.rm = TRUE)/sqrt(length(time)),
            group_sd   = sd(Values,na.rm = TRUE),
            SEM_upper = max(group_SEM)+group_mean,
            SEM_lower = min(group_SEM)+group_mean,
            sd_upper = max(group_sd)+group_mean,
            sd_lower = min(group_sd)+group_mean)
```
    

### b) Line plot with discrete error bars showing SD

Use `ggplot2` to create a line graph to plot the mean, with discrete error bars representing the standard deviation.

+ Map the data to time (x-axis) and mean (y-axis), and use color to indicate the treatment.
+ Use a line for the geom layer.
+ Add the error bars using `geom_errorbar()` (look this up to see how to add the upper and lower boundaries). Don't forget to map the color to the treatment.
+ Add axis labels for the time (in seconds) and the values (as "Relative fluorescence").

```{r}
library(ggplot2)
ggplot(data = c.elegans_dplyr, aes(x = time, y = group_mean, col = TREATMENT)) +
  geom_line() +
  geom_errorbar(aes(x = time, ymin = sd_lower, ymax = sd_upper)) + 
  labs(x = "Time in seconds", y = "Relative Fluorescence")
```


### c) Line plot with shaded regions showing SEM

Use `ggplot2` to create a line graph to plot the mean, with a shaded area representing the standard error of the mean. This plot will use almost the exact same code as the last graph, except you will use `geom_ribbon()` to display the upper and lower bounds of the SEM.
      
```{r}
library(ggplot2)
ggplot(data = c.elegans_dplyr, aes(x = time, y = group_mean, col = TREATMENT)) +
  geom_line() +
  geom_ribbon(aes(x = time, ymin = SEM_lower, ymax = SEM_upper)) + 
  labs(x = "Time in seconds", y = "Relative Fluorescence")
```


### d) Which representation do you think is most effective for these data? What would you choose to do?
```{r}
# I prefer the second representation as it more continuously represents the range that the mean frequency distribution and the mean error distribution covers, so I can draw a more concrete conclusion as to which time interval yields a significant difference in relative fluorescence.
```

