---
title: 'Homework 5: Contingency Tests'
subtitle: "XDASI Fall 2021"
author: "Dylan Fitzmaurice"
date: "Due: 10/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE, eval=TRUE)
```


# Association between obesity and genetic polymorphisms

## Q0: Load BMI Data

This is a published data set containing measurements of Body Mass Index (BMI) and single-nucleotide polymorphisms (SNPs) based on genotyping of 79 individuals. The data file provided contains the BMI (numeric) and the presence or absence of 7 SNPs (a factor, encoded as Y or N) for each individual.

Our goal in this study is to determine whether any of the SNPs are associated with Obesity.

#### a) Load the `HW6_SNP_Data.tsv` file into a dataframe called `hw6.data`.

```{r}
hw6.data = read.table("/Users/dylanfitzmaurice/R/Biostats/Week_6/HW6_SNP_Data.tsv")
```

#### b) Briefly describe the structure of this observational data set. 

What measurements have been taken, and what is the data type of those measurements?

```{r}
# your answer here
# BMI and 7 SNPs. The data type of BMI is numeric, while SNPs are logical (Y/N). 
```


## Q1: $\chi^2$ Test

#### a) Add an column for `Overweight` status

The column should have binary values (e.g. TRUE/FALSE) according to whether the BMI for each individual is 25 or above, or under 25.


```{r}
hw6.data$V9 = hw6.data$V1 >= 25
hw6.data$V9[1] = "Overweight"

```

#### b) How many individuals are overweight by this definition? How many are not overweight?

```{r}
table(hw6.data$V9)["TRUE"] #TRUE
table(hw6.data$V9)["FALSE"] #FALSE
```

#### c) Generate a contingency table for Overweight and SNP1.

```{r}
ow_successes = table(hw6.data$V9)["TRUE"]
ow_failures = table(hw6.data$V9)["FALSE"]

SNP1_successes = table(hw6.data$V2)["Y"]
SNP1_failures = table(hw6.data$V2)["N"]

# arrange as a contingency table
ow_snp1_table <- rbind(c(ow_successes, SNP1_successes),
                      c(ow_failures, SNP1_failures))
colnames(ow_snp1_table) = c("Overweight","SNP1")
rownames(ow_snp1_table) = c("TRUE", "FALSE")
ow_snp1_table
```

#### c) Generate a table of expected counts.

Make a table showing the **expected** number of individuals in each cell for the Overweight / SNP1 contingency table, assuming independence:

$$E_{A{\cap}B} = Pr_A * Pr_B * Total = n_A * n_B/ N$$

```{r}
N = sum(ow_snp1_table)
T_total = sum(ow_snp1_table[1,])
F_total = sum(ow_snp1_table[2,])
OW_total = sum(ow_snp1_table[,1])
SNP1_total = sum(ow_snp1_table[,2])

Expected = c( T_total * OW_total  / N,
              F_total  * OW_total  / N,
              T_total * SNP1_total / N,
              F_total  * SNP1_total / N)

ExpectedValues = matrix(Expected,nrow=2,ncol=2)
ExpectedValues
```

#### d) $\chi^2$ statistic

Calculate the $\chi^2$ test statistic for association between Obesity and SNP1 based on the observed and expected counts.

```{r}

chisq = sum((ow_snp1_table[1:2,1:2] - ExpectedValues)^2 / ExpectedValues )
chisq


```


#### e) Compute a $p$-value for the test statistic

Use the CDF for the $\chi^2$ distribution to compute the significance of the test statistic.

```{r}
pchisq(chisq, df = 1, lower.tail = F)
```

#### f) Confirm your results using R's Chi-squared test function

*Note: Turn Yates' correction on or off depending on whether you used this for the manual calculation above.*

```{r}
chisq.test(ow_snp1_table,correct = FALSE)

```

#### g) What is the null hypothesis for this chi-squared test?

```{r eval=FALSE}
# your answer here
#There is no association between SNP1 and overwieghtness 
```

#### h) Interpret the result of the Chi-squared test between the Overweight trait and SNP1.

```{r eval=FALSE}
# your answer here
#SNP1 is not associated with being overweight
```


## Q2: Fisher's exact test

#### A) Describe the difference between Fisher's Exact Test and the Chi-Squared test.

```{r eval=FALSE}
# your answer here
# Fisher's exact test uses the hypergeometric distribution to provide an exact 
# p-value, while Chi-squared is an approximation that assumes the sample size 
# is large

#### b) Test the relationship between the Overweight trait and SNP1 with a Fisher's Exact Test.

```{r}
fisher.test(ow_snp1_table)
```

#### c) What is the null hypothesis for this Fisher's Exact Test?

```{r eval=FALSE}
# your answer here
# That being overweight and having SNP1 are independent
```


#### d) Interpret the result of your Fisher's Exact test. Compare the p-values to the chi-squared test.

```{r eval=FALSE}
# your answer here
#The results are close but different because continuity was not corrected for 
#in the chi-squared test.
```


## Q3: Measures of Association

#### a) How are the Relative Risk and Odds Ratio defined?

```{r}
# your answer here
# RR: the probability of an undesired outcome in the treatment group divided by 
# the probability of the same outcome in a control group.
  
# OR:the probability of success divided by the probability of failure

```

#### b) Calculate the RR for SNP1 and Obesity.

```{r}
a=ow_snp1_table[1,1]
b=ow_snp1_table[1,2]
c=ow_snp1_table[2,1]
d=ow_snp1_table[2,2]

p1 = a/(a+c)
p2 = b/(b+d)

RR = p1 / p2
RR
```

#### c) Confirm your results using the `riskratio` function in the `epitools` package.

*Your use of Yates' correction should match whatever you did above.*

```{r}
library(epitools)
riskratio(ow_snp1_table)
```

#### d) Calculate the OR for SNP1 and Obesity manually.

```{r}
OR = (a/c) / (b/d)
OR
```

#### d) Confirm your results using the `oddsratio` function in the `epitools` package.

Note: Use the `wald` method for maximum likelihood estimation.

*Your use of Yates' correction should match whatever you did above.*

```{r}
oddsratio.wald(ow_snp1_table)
```

#### e) How did the result of your Chi-squared and Fisher's tests line up with the report from the odds ratio function?

```{r}
# your answer here
#Not the same value, because chi-square continuity correction is set to false
```


## Q4: Check the other SNPs to see which SNP is most associated with obesity.

#### a) Use Fisher's Exact Test to check for this.

*Hint: You may want to use a loop to do this. Recall that you can extract the p-value from the test report using the `$p.value` notation.*

```{r}

pvalue = numeric(6)
for (i in 3:8) {
# arrange as a contingency table
ow_snp3_to_7_table <- rbind(c(ow_successes, table(hw6.data[2:80,i])["Y"]),
                       c(ow_failures, table(hw6.data[2:80,i])["N"]))
colnames(ow_snp3_to_7_table) = c("Overweight","SNP2")
rownames(ow_snp3_to_7_table) = c("TRUE", "FALSE")

pvalue[i-2] = fisher.test(ow_snp3_to_7_table)$p.value}

pvalue

```

#### b) Are any of the SNPs significantly asociated with BMI?

```{r eval=FALSE}
# your answer here
#Yes, SNP4 and 6. 
```
