---
title: 'Homework 6: Contingency Tests'
subtitle: "XDASI Fall 2021"
author: "Ken Tanaka"
date: "Due: 10/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE, eval=FALSE)
```


# Association between obesity and genetic polymorphisms

## Q0: Load BMI Data

This is a published data set containing measurements of Body Mass Index (BMI) and single-nucleotide polymorphisms (SNPs) based on genotyping of 79 individuals. The data file provided contains the BMI (numeric) and the presence or absence of 7 SNPs (a factor, encoded as Y or N) for each individual.

Our goal in this study is to determine whether any of the SNPs are associated with Obesity.

#### a) Load the `HW6_SNP_Data.tsv` file into a dataframe called `hw6.data`.

```{r}

hw6.data <- read.table("data/HW6_SNP_Data.tsv", header = T, sep = "\t", stringsAsFactors = F)
#hw6.data
head(hw6.data)

```

#### b) Briefly describe the structure of this observational data set. 

What measurements have been taken, and what is the data type of those measurements?

```{r}

str(hw6.data)
summary(hw6.data)
subset(hw6.data, is.na(hw6.data$BMI))

library(dplyr)
summarise(group_by(hw6.data,SNP1,SNP2,SNP3,SNP4,SNP5,SNP6,SNP7), n = n())

library(gmodels)
ci(hw6.data$BMI)

#your answer here
"From summary(), it was confirmed that N/A data was not there.  The minimum value of BMI was 18.65, the maximum value of BMI was 46.66, and the mean was 26.30. Furthermore, the lower limit of the 95% confidence interval was 24.93 and the upper limit was 27.66." 
```


## Q1: $\chi^2$ Test

#### a) Add an column for `Overweight` status

The column should have binary values (e.g. TRUE/FALSE) according to whether the BMI for each individual is 25 or above, or under 25.


```{r}
Overweight = c()
recnum = length(hw6.data$BMI)
for (i in 1:recnum) {
  if (hw6.data$BMI[i] >= 25) {
    Overweight <- c(Overweight,  T)
  } else {
    Overweight <- c(Overweight,  F)
  }
}
Overweight
hw6.overweight <- cbind(hw6.data, Overweight)
hw6.overweight
str(hw6.overweight)
```

#### b) How many individuals are overweight by this definition? How many are not overweight?

```{r}
summarise(group_by(hw6.overweight, Overweight), n=n())
```

#### c) Generate a contingency table for Overweight and SNP1.

```{r}
hw6.SNP1.OWt <- summarise(group_by(hw6.overweight, SNP1, Overweight), n=n())
hw6.SNP1.OWt
#str(hw6.SNP1.OWt)
SNP1N_OwtF = hw6.SNP1.OWt$n[1]
SNP1N_OwtF
SNP1N_OwtT = hw6.SNP1.OWt$n[2]
SNP1N_OwtT
SNP1Y_OwtF = hw6.SNP1.OWt$n[3]
SNP1Y_OwtF
SNP1Y_OwtT = hw6.SNP1.OWt$n[4]
SNP1Y_OwtT
# Observed = rbind(c(SNP1N_OwtT, SNP1Y_OwtT),
#                 c(SNP1N_OwtF, SNP1Y_OwtF))

Observed = rbind(c(SNP1Y_OwtT, SNP1Y_OwtF),
                 c(SNP1N_OwtT, SNP1N_OwtF))
colnames(Observed) = c("Overweight","Healthy")
rownames(Observed) = c("SNP1Y", "SNP1N")
Observed
```

#### c) Generate a table of expected counts.

Make a table showing the **expected** number of individuals in each cell for the Overweight / SNP1 contingency table, assuming independence:

$$E_{A{\cap}B} = Pr_A * Pr_B * Total = n_A * n_B/ N$$

```{r}
#SNP1N_OwtF
#SNP1N_OwtT
#SNP1Y_OwtF
#SNP1Y_OwtT
SNP1N_total = SNP1N_OwtF + SNP1N_OwtT
SNP1N_total
SNP1Y_total = SNP1Y_OwtF + SNP1Y_OwtT
SNP1Y_total
OwtF_total = SNP1N_OwtF + SNP1Y_OwtF
OwtF_total
OwtT_total = SNP1N_OwtT + SNP1Y_OwtT
OwtT_total
N = SNP1N_OwtF + SNP1N_OwtT + SNP1Y_OwtF + SNP1Y_OwtT

 SNP1N_OwtT_expected = SNP1N_total * OwtT_total / N
 SNP1Y_OwtT_expected = SNP1Y_total * OwtT_total / N
 SNP1N_OwtF_expected = SNP1N_total * OwtF_total / N
 SNP1Y_OwtF_expected = SNP1Y_total * OwtF_total / N

# Expected = rbind(c(SNP1N_OwtT_expected, SNP1Y_OwtT_expected),
#                 c(SNP1N_OwtF_expected, SNP1Y_OwtF_expected))
# rownames(Expected) = c("Overweight","Healthy")
# colnames(Expected) = c("SNP1N", "SNP1Y")
 
 
 
Expected = rbind(c(SNP1Y_OwtT_expected, SNP1Y_OwtF_expected),
                 c(SNP1N_OwtT_expected, SNP1N_OwtF_expected))
colnames(Observed) = c("Overweight","Healthy")
rownames(Observed) = c("SNP1Y", "SNP1N")
 
 
 
 
Expected

```

#### d) $\chi^2$ statistic

Calculate the $\chi^2$ test statistic for association between Obesity and SNP1 based on the observed and expected counts.

```{r}
chisq = sum( abs(Observed - Expected)^2 / Expected)
chisq

```


#### e) Compute a $p$-value for the test statistic

Use the CDF for the $\chi^2$ distribution to compute the significance of the test statistic.

```{r}

pchisq(chisq, df = 1, lower.tail = F)
```

#### f) Confirm your results using R's Chi-squared test function

*Note: Turn Yates' correction on or off depending on whether you used this for the manual calculation above.*

```{r}
chisq.test(Observed)
```

#### g) What is the null hypothesis for this chi-squared test?

```{r eval=FALSE}
# your answer here
"Since the p-value of this chi-squared test was 1, it is not possible to reject the null hypothesis."

```

#### h) Interpret the result of the Chi-squared test between the Overweight trait and SNP1.

```{r eval=FALSE}
# your answer here
"Since the p-value of this chi-squared test was 1, it is not possible to reject the null hypothesis. Therefore, it can be said that there are relationship with Overweight trait and SNP1."

```


## Q2: Fisher's exact test

#### A) Describe the difference between Fisher's Exact Test and the Chi-Squared test.

```{r eval=FALSE}
# your answer here
"The chi-squared test applies an approximation assuming the sample is large, while the Fisher's exact test runs an exact procedure especially for small-sized samples."
```

#### b) Test the relationship between the Overweight trait and SNP1 with a Fisher's Exact Test.

```{r}
Observed
fisher.test(Observed)
```

#### c) What is the null hypothesis for this Fisher's Exact Test?

```{r eval=FALSE}
# your answer here
"The null hypothesis for this Fisher's exact test is that Overweight and SNP1 classifications are not different."
 
```


#### d) Interpret the result of your Fisher's Exact test. Compare the p-values to the chi-squared test.

```{r eval=FALSE}
fisher.test(Observed)$p.value

# Since the CDF for the chi-squared distribution is 0.8719426 (Please see Q1.e) and the p-value was 1 computed above, it is not possible to reject the null hypothesis. Therefore, the Overweight and SNP1 classifications are not different.   

```


## Q3: Measures of Association

#### a) How are the Relative Risk and Odds Ratio defined?

```{r}
# your answer here
"The Relative risk is defined as the ratio of the proportion of subjects with a focal outcome in one group vs. the other group. An odds ratio (OR) is a measure of association between an exposure and an outcome. The OR represents the odds that an outcome will occur given a particular exposure, compared to the odds of the outcome occurring in the absence of that exposure."
```

#### b) Calculate the RR for SNP1 and Obesity.

```{r}
row3 <- c(SNP1Y_OwtT + SNP1N_OwtT, SNP1Y_OwtF + SNP1N_OwtF)

Observed
row3col2 <- rbind(Observed, row3)
row3col2

col3 <- c(SNP1Y_OwtT + SNP1Y_OwtF, SNP1N_OwtT + SNP1N_OwtF, SNP1Y_OwtT + SNP1Y_OwtF + SNP1N_OwtT + SNP1N_OwtF)
col3

row3col3 <- cbind(row3col2, col3)
row3col3

str(row3col3)

# Calculate the Relative Ratio for SNP1 and Obesity. 

row3col3[1,1]

RR = (row3col3[1,1]/row3col3[1,3])/(row3col3[2,1]/row3col3[2,3])
RR


```

#### c) Confirm your results using the `riskratio` function in the `epitools` package.

*Your use of Yates' correction should match whatever you did above.*

```{r}
#install.packages("epitools")
library("epitools")
riskratio(Observed)
riskratio(Observed, correction = T, rev = "both")

# The values were matched with Yates' correction. 

#riskratio(Observed, rev="both")


```

#### d) Calculate the OR for SNP1 and Obesity manually.

```{r}

# Calculate the Odds Ratio for SNP1 and Obesity. 

OR = (row3col3[1,1]/row3col3[2,2])/(row3col3[1,2]/row3col3[2,1])
OR


```

#### d) Confirm your results using the `oddsratio` function in the `epitools` package.

Note: Use the `wald` method for maximum likelihood estimation.

*Your use of Yates' correction should match whatever you did above.*

```{r}
#install.packages("epitools")
library("epitools")
oddsratio(Observed)
oddsratio(Observed, correction = T, rev = "both")

# The values were matched with Yates' correction. 

```

#### e) How did the result of your Chi-squared and Fisher's tests line up with the report from the odds ratio function?

```{r}
# your answer here
"The results of the Chi-squared and Fisher's exact tests were consistent with the report from the odds ratio function."

```


## Q4: Check the other SNPs to see which SNP is most associated with obesity.

#### a) Use Fisher's Exact Test to check for this.

*Hint: You may want to use a loop to do this. Recall that you can extract the p-value from the test report using the `$p.value` notation.*

```{r}
snpowt <- function(hw6.SNP.Owt){
 
 print(hw6.SNP.Owt)
 
 SNPN_OwtF = hw6.SNP.Owt$n[1]
 #print(SNPN_OwtF)
 SNPN_OwtT = hw6.SNP.Owt$n[2]
 #print(SNPN_OwtT)
 SNPY_OwtF = hw6.SNP.Owt$n[3]
 #print(SNPY_OwtF)
 SNPY_OwtT = hw6.SNP.Owt$n[4]
 #print(SNPY_OwtT)

 Observed_SNP = rbind(c(SNPY_OwtT, SNPY_OwtF),
                     c(SNPN_OwtT, SNPN_OwtF))
 colnames(Observed_SNP) = c("Overweight","Healthy")
 rownames(Observed_SNP) = c("SNPY", "SNPN")
 #str(Observed_SNP)
 print(colnames(hw6.SNP.Owt))
 print(Observed_SNP)
 
 obs_pval <- fisher.test(Observed_SNP)$p.value
 print(obs_pval)
}

#str(hw6.overweight)
coln <- colnames(hw6.overweight)
SNPN = coln[grep("SNP", coln)]
SNPN
snplen = length(SNPN)
hw6.SNP.OWt <- data.frame()
SNP <- ""
#Observed_SNP <- NULL
for (i in 1:snplen) {
 SNP = SNPN[i]
 eval(parse(text = paste("hw6.SNP.Owt <- summarise(group_by(hw6.overweight,", SNP ,", Overweight), n=n())") ))
 #str(hw6.SNP.Owt)
 snpowt(hw6.SNP.Owt)
}

```

#### b) Are any of the SNPs significantly asociated with BMI?

```{r eval=FALSE}
# your answer here
"Since the p-value of SNP3 was 0.2502351 which was the lowest p-value among the seven SNPs in the above, the SNP3 is the most significantly associated with Overweight. Therefore, the SNP3 is the mostly significantly associated with BMI."

```
