---
title: 'Homework 5: Binomial Distribution'
subtitle: "XDASI Fall 2021"
author: "Jacob Bernoulli [Omar Qassab]"
date: "Due: 10/8/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
```

### Note

For these exercises, please note the following:

+ You should review the addition and multiplication rules for the joint probability of independent and mutually exclusive events.
+ You will sometimes need to use the `choose()` function, which computes the binomial coefficient for you. Look up the documentation for this function to find out how to use it.
+ You will also need to learn how to use the `binom()` family of functions.


## 1. Autosomal recessive disorder

Many disease genes have autosomal recessive phenotypes. Below, assume that both parents are heterozygous for to alleles, A and a, and the aa genotype displays the disease phenotype.

a. What are the individual probabilities of each allele combination among the progeny?

```{r}
P_AA = 1/4
P_Aa = 2/4
P_aa = 1/4

paste("The probabilites are: ", P_AA,"Homozygous Dominant,", P_Aa, "Heterozygote", P_aa, "Homozygote recessive")
```

b. What is the probability that 1 out of 3 children will have the disease?

```{r}
P_one_disease= choose(3,1) * P_aa * ((1-P_aa)^2)

paste("The probability of one child having the disease is equal to", P_one_disease)
```

c. What is the probability that all 3 children will be carriers (i.e. they have the allele but do not have the disease)?

```{r}
P_all_carrier=P_Aa^3

paste("The probability that all 3 children are carriers is equal to", P_all_carrier)
```

d. What is the probability that at least one child will be a carrier (i.e. they have the a allele but do not have the disease)?

```{r}
P_one_carrier=choose(3,1)*P_Aa *((1-P_Aa)^2)
P_two_carrier=choose(3,2) * (P_Aa^2) * (1-P_Aa)
P_three_carrier=P_Aa^3
P_atleast_one_carrier = P_one_carrier + P_two_carrier + P_three_carrier

paste("The probability that one child will be a carrier is equal to", P_one_carrier)
paste("The probability that at least one child will be a carrier is equal to", P_atleast_one_carrier)
```

e. Now imagine that the disease is recessive lethal, so children with the disease will not survive gestation. What is the probability that at least one child will be a carrier?

```{r}
P_one_carrier_lethal = choose(3,1) * 0.66 * (0.33^2)
P_two_carrier_lethal= choose(3,2) * (0.66^2) * 0.33
P_three_carrier_lethal = 0.66^3
P_atleast_one_carrier_lethal= P_one_carrier_lethal + P_two_carrier_lethal + P_three_carrier_lethal

paste("The probability that one child will be a carrier equals", P_one_carrier_lethal)
paste("The probability that at least one child will be a carrier equals", P_atleast_one_carrier_lethal)
```

f. What is the probability that all 3 children will be carriers?

```{r}
P_all_carrier_lethal = 0.66^3

paste("The probability that all three children will be carriers equals", P_all_carrier_lethal)
```


## 2. Wnt signaling

Proliferation of embryonic stem cells is important for early development and is promoted by Wnt signaling, which promotes cell cycle progression through the transcriptional activator beta-catenin. In a study of the developing mouse, it was found that only 11% of cells in a region of the brain that is rich in neuronal progenitors was responsive to Wnt (even though Wnt is required for stem cell expansion).

Here we will examine different possible outcomes we could observe using FACS to sort cells from this region that are labeled with a fluorescent antibody against the Wnt receptor. Assume you are sampling from a relatively large number of cells (>10,000). 

Below, we will adopt the following conventions:

+ **W** = Wnt-responsive cells (cells with the receptor and are fluorescent)
+ **I** = Wnt-insensitive cells (no receptor, no fluorescence)

a. Use **only the addition and multiplication rules** for probabilities to find the probability of observing each of the outcomes below among a sample of 6 cells. 

*Hint: would you consider this sampling with or without replacement, if you are sampling <100 cells?*

```{r collapse=TRUE}
## Here we are sampling without replacement

# probability of different cell types
pW = 0.11
pI = 0.89

# 1. A sequence of WWIIII
pW*pW*pI*pI*pI*pI
# 2. A sequence of IIWWII
pI*pI*pW*pW*pI*pI
```

b. How many ***ways*** (combinations) are there to get four Wnt-insensitive cells out of 6 total? This is the same as the number of ways to get two Wnt-responsive cells out of 6 (why?).

```{r}
paste("There are",choose(6,4),"to get four Wnt-insensitive cells out of 6 total.", "This is the same as getting two Wnt-responsive cells out of 6. This is because four insensitive cells means we have to have two sensitive ones") 

```

c. What is the total ***probability*** of getting **exactly** 4 I cells and 2 W cells in your sample, taking into account all the different ways that this outcome can be obtained?

+ First, answer this using `choose()` and the individual probabilities (according to the formula for binomial proportions).
+ Then, use one of the `binom()` functions to compute the same result.

For the rest of this question, we will consider observing a Wnt-responsive cell in a single Bernoulli trial to be a ***"success"***.

```{r}
# with choose
choose(6,4)*(pI^4)*(pW^2)

# with a binom function
dbinom(4,6,pI)
```

d. What would be the ***probability*** of getting **2 or less** W cells in a sample (again considering "W" to be a "success")?

+ First, answer this using `choose()` and the individual probabilities.
+ Then, use one of the `binom()` functions to compute the same result.

```{r}
# with choose and probabilities
p_two_or_less=0


#For loop makes it easier
for (i in c(0:2)) {
  p_two_or_less = p_two_or_less + 
    (choose(6,i)*(pW^i)*(pI^(6-i) ))
  
}

#This is what I want happening in my for loop
(choose(6,2) * (pW^2) *(pI^4)) +
  (choose(6,1) * (pW^1) *(pI^5)) +
  (choose(6,0) * (pW^0) *(pI^6))

paste("The probability of getting 2 or less W in a sample is", p_two_or_less)


# with a binom function
pbinom(2,6,0.11)
```


e. How many ***ways*** are there to get **at least one** (i.e. 1 or more) W cell in a sample of 6 cells? 

*Hint: `choose()` can take a vector, and you can use `sum()` to add up the combinations, so you don't really have to add all of the individual terms up by hand.*

```{r}
# example
# choose(6,3) + choose(6,4)
# sum(choose(6,3:4))

# your answer here
sum((choose(6,1:6)))

```

e. What is the ***probability*** of getting 6 I's and no W's?

```{r}
pI^6
```

f. What would be the ***probability*** of getting **at least one** (i.e. 1 or more) W cell in a sample of 6 cells? You can calculate this at least 4 equivalent ways given the tools you have!

***Method 1):*** Use only addition and multiplication rules for probabilities

*Hint: This could be 1 or more W's with any combination of W's or I's. This is easy to solve when you remember that it is the same as Pr(NOT IIIIII) !*  :)
 
```{r}

1-(pI^6)

```

***Method 2):*** Use the binomial formula with `choose()`, pW, and pI

*Again, remember that this is the same as 1 - Pr(0 W's)*

```{r}
p_one_or_more = 0

for (i in c(1:6)) {
  p_one_or_more = p_one_or_more + 
    (choose(6,i)*(pW^i)*(pI^(6-i) ))
  
}


paste("The probability of getting 1 or more W in a sample is", p_one_or_more)

#or

1-(choose(6,0)*(pW^0)*(pI^6))


```

***Method 3):*** Use the `dbinom()` function

*Hint: `dbinom()` also takes a vector, so you can use the `sum()` function to add up all of the discrete probabilities for W>0.*

```{r}
sum(dbinom(c(1:6),6,pW))
```

***Method 4:*** Use the `pbinom()` function

```{r}
pbinom(0,6,pW,lower.tail = F)
```

---

## 3. Ecological influence of pollution

During the industrial revolution in England, London become covered in a lot of black soot from burning coal. Five years beforehand, the proportion of white moths that could be found in London was 87%.[^1] However black moths gained a survival advantage as the air became more polluted. 

If you were to sample moths in London 5 years into the industrial revolution, and you found that 35 out of 50 moths were white, how likely would that be if the population had remained the same?

a. Use the `choose()` function for the binomial coefficient and the addition and multiplication rules for probabilities to find how likely it would be to catch ***exactly*** 35/50 white moths if P(white) remained unchanged.

```{r}
p = 0.87
q = 0.13

# probability of 35/50 white given p
choose(50,35)*(p^35)*(q^15)

```


b. Use one of the R `binom()` functions to compute this probability.

```{r}
dbinom(35,50,p)
```


c. Really what you are probably more interested in is finding out how likely it is that you only found _**no more than**_ 35 in total. Given the null distribution, what is the total probability of this event? Does this result seem significant?

```{r}
# how likely is it to see only 35/50 white moths?
binom.test(35,50,p)

# is this significant?
#The p-value is equal to 0.001285. This means that we reject the null, and the distribution of moths is not the same as in our binomial distribution
```


d. Confidence interval

The `binom` package has a method called `binom.confint()` that will compute confidence intervals for the binomial distribution using a large number of different methods. Download this package, install the library, and use it to compute a 95%CI for the true population parameter given the data. 

What is your best estimate for P(white), and what are the mean and 95% CI for the number of white moths you'd expect to see in your sample of 50 post-industrial moths?

```{r}
#install.packages("binom")  # do this once
library(binom)
moth.ci = binom.confint(35,50)
moth.ci

# P(white) and 95% CI for 50 moths (multiply probabilities by sample size)
# below we use the concatenate function to print out mixed text and variables
# just plug in the relevant numbers from moth.ci below
cat("Mean estimate = ", 0.70  , 
    "; 95%CI = ",       0.5617317 * 50 ,
    "-",                0.8097293 * 50 )
```


e. Use R's `par()` and `plot()` function to draw side-by-side plots showing the ideal binomial **PDF and CDF** for the null distribution with n=50 and p=P(white)=0.11.

+ For the PDF, use "histogram-like vertical lines".
+ For the CDF, use "stair-steps".

*Hint: check the documentation to find the plot `type` parameter for a histogram or a step type.*

```{r}
## Option 1: just make a quick plot all in one go
par(mfrow=c(1,2))

# create a vector for the x-axis ranging from 0-50
x.range = 0:50 

# plot x on the x-axis and distributions on the y-axis
plot(x= x.range , y= dbinom(c(0:50),50,0.11), type= "h" ,
     main="PDF",
     xlab="expected # of white moths", ylab="point probability")
plot(x=x.range , y= pbinom(c(0:50),50,0.11) , type= "s" ,
     main="CDF",
     xlab="expected # of white moths", ylab="total probability")


# ## Option 2: step by step
# 
# # first specify the expected proportion (probability) of successes
# pW = 0.11
#   
# # create a vector for the x-axis ranging from 0-50
# x.range = 0:50 
# 
# # PDF
# dbinom.W = dbinom(c(0:50),50,pW)
# 
# # CDF
# pbinom.W = pbinom(c(0:50),50,pW)
# 
# # make the plots (as above)
# 
# par(mfrow=c(1,2))
# plot(x= x.range , y= dbinom.W, type= "h" ,
#      main="PDF",
#      xlab="expected # of white moths", ylab="point probability")
# plot(x=x.range , y= pbinom.W , type= "s" ,
#      main="CDF",
#      xlab="expected # of white moths", ylab="total probability")

```

f. What kind of data structure would you need to create in order to use the `hist()` function or `ggplot` instead?

```{r}
# You would need a data frame
```

---

## Extra credit: write a CDF function

1. Use a loop to compute the the cumulative probability of seeing between **0** and **x** moths in a sample of size **n** by adding up all the discrete probabilities for a discrete variable. 
2. Wrap this in a function.
3. Then, call your function to find the total probability of seeing no more than 35/50 white moths, given the null distribution with P(W) = 0.11 or P(W) = 0.87. *Note that for both of the examples here, we considered W to be a success, we just changed p = P(W).*

```{r}
# define the function
cdf <- function(x,n,p) {
  
  # x is the number of successes
  # n is the number of trials
  # p is the probability of success
  
  #compute failure probability
  q=1-p
  
  #set up initial probability variable
  result = 0
  
  #for loop computing CDF
  for (i in 0:x) { 
    result= (choose(n,i) * (p^i) * (q^(n-i) ) ) + result
      
  }
  
  # Return the results
  return( result )
}

# call the function for two different probability values
cdf(35,50,0.11)
cdf(35,50,0.87)
```

---

## Extra credit: Make a histogram of the ideal PDF using ggplot2

***Before you start***, go to this website: <a href="https://www.stat.berkeley.edu/~stark/Java/Html/BinHist.htm" target="blank">**SticiGui**</a>. 

+ If you enter (n=50, p=0.11), you will see approximately what the probability distribution for Question 1 should look like. For Question 2, just change p=0.87.

**Follow the example from recitation** to draw  plot of the ideal binomial PDF we would expect to get for **x** "successes" in a sample of size **n**, given any probability of "success" **p**. 

Let's also pretend we replicated this on 20 different samples of size 50, so we will be plotting the number of replicates in which we would expect to find 1,2,3,..,49,50 W"successes" out of 50.

Then, follow these steps:

1. Generate a vector of the expected ***probability distribution*** for the number of "successes" observed out of 50 random trials based on an ideal binomial distribution.
  + For P(success), use either P(W) = 0.11 or P(W) = 0.87.
2. Multiply this by the number of replicates to get the **frequency distribution** and round off to the nearest integer.
3. Convert this to a "simple" vector of counts.
4. Finally, make a histogram of these data. This should result in graphs that show the expected number of times that X out of 50 successes were observed in 20 replicate experiments.

```{r}
# create variables for binomial sampling and initialize them with your choice of p
sample.size = 50
p.success = 0.11

# probability distribution
ideal.sample = dbinom(0:sample.size, sample.size, p.success)
ideal.sample

# convert probabilities to frequencies
# (multiply by number of replicate experiments)
replicates = 20
ideal.freq = round(ideal.sample*replicates,0)
ideal.freq

# convert frequencies to count vector of observed proportion of W cells
# e.g. for freq.W = 0 0 1 2 3 4 3 3 2 1 1 0 0 ... we have
# 0 reps with 0 W, 0 reps w/ 1 W, 1 rep w/ 2 W, 2 reps w/ 3W, 3 reps w/ 4 W, etc.
# so we convert this to: counts.W = c(2,3,3,4,4,4,5,5,5,5,6,6,6,7,7,7,8,8,9,10)
ideal.counts = c() # create an empty vector

for (i in 0:sample.size) {
  ideal.counts=append(ideal.counts,rep(i,ideal.freq[i+1]))
    
}

ideal.counts

bino_data <- data.frame("successes" = ideal.counts)

# plot all frequencies
ggplot(data = bino_data, mapping = aes(x = successes)) +
  # center columns on integers
  geom_histogram(breaks = seq(from = -0.5, to = sample.size+0.5, by = 1),
                 col = "midnightblue", fill="cornflowerblue") +
  # declutter labels
  scale_x_continuous(name="Expected number of successes",
                     breaks=seq(0,sample.size,5),
                     limits = c(-1,sample.size)) +
  ylab("Frequency") +
  ggtitle(paste0("Ideal binomial frequency distribution for 20 replicates (n=",
  sample.size, ", p=", p.success, ")"))
```

Now answer the following questions:
```{r}
# What is the most frequent number of W observed out of a sample size of 50?
#5

# How does this gibe with the given probability P(W)?
#It's about the same

# What is the total height of all the bars if you sum them all up?
#20

# What does this number correspond to?
#The number of replicates

# Where would you expect to the peak of the distribution to be if P(W)  = 0.35?
#It would be around 17-18.
```

[^1]: Disclaimer: I made this up, but based on a true story (see https://en.wikipedia.org/wiki/Peppered_moth_evolution).