---
title: 'In-class Exercises: Binomial Distribution'
author: "Kris Gunsalus"
date: "9/29/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Bernoulli distribution

The frequency of a minor allele at the CFTR gene was found to be 0.17. What is the probability that any allele chosen at random from the population will NOT be that same allele?

$$ P(X=0) = f(0) = (0.17)^0(1-0.17)^{1} = 0.83$$


## Binomial distribution (Example 1)

During the industrial revolution in England, London become covered in a lot of black soot from burning coal. Five years beforehand, the proportion of white moths that could be found in London was 87%.[^1] However black moths gained a survival advantage as the air became more polluted. 

If you were to sample moths in London 25 years into the industrial revolution, and you found that 35 out of 50 moths were white, how likely would that be if the population had remained the same?

You can compute the probability of finding _**exactly**_ 35/50: (Answer: 0.088%)

$$ P(X = 35) = {50 \choose 35} 0.87^{35}(1-.87)^{50-35} $$

Which R function should we use to compute this probability, and what is it?

```{r}
# PDF
dbinom(35,size=50,prob=0.87)

# CDF
pbinom(35,size=50,prob=0.87) - pbinom(34,size=50,prob=0.87) 
```

Really what you are probably more interested in is finding out how likely it is that you only found _**no more than**_ 35 in total. To do this, you'd have to add up all the probabilities of getting 0, 1, 2, ... 35 white moths. This is given by the binomial CDF:

$$ P(X \le 35) = \sum_{x_i \le 35}f(x_i)
= \sum_{x_i \le 35}{50 \choose x_i}0.87^{x_i}(1-0.87)^{50-x_i} $$

Wow, that looks nasty! You could use a loop to compute this in R:
```{r}
cdf <- function(x,n,p) {
  result = numeric()
  for (i in 0:x) {
    f = n-i
    q = 1-p
    result[i] = choose(n,i)*(p^i)*(q^f)
  }
  return(sum(result))
}
cdf(35,50,0.87)
```

... but the R function for the binomial CDF is so much simpler! It is:
```{r}
pbinom(35,50,0.87) # cumulative probability for x between 0 and 35
```

It's hard to tell from just one sample what the true proportion really is ... we will talk a lot more about this soon! For now, we can estimate how likely it would be that you would find, say, somewhere between 30-40 white moths, vs. 40-50, if the population had not changed:
```{r}
sum(dbinom(30:40,50,0.87))  # very unlikely!
sum(dbinom(40:50,50,0.87))  # much more likely
```

To find the probability of getting 40 or more white moths if the population had stayed the same, i.e. $P(X \ge 40)$, you could also use the CDF. You'd have to add `lower.tail = FALSE` and use 39 instead of 40, or else you would be asking for $P(X > 40)$:
```{r}
pbinom(39,50,0.87,lower.tail=FALSE)
```

We can visualize the density and the CDF by plotting them, which makes it a lot easier to understand the probabilities we computed above:
```{r}
par(mfrow=c(1,2))
x = 0:50
plot(x,dbinom(x,size=50,prob=0.87),type="h",main="PDF",
     xlab="number of successes",ylab="point probability")
plot(x,pbinom(x,size=50,prob=0.87),type="h",main="CDF",
     xlab="number of successes",ylab="total probability")
```


## Binomial distribution  (example 2)

Consider a genome in which the four bases $A,C,G,T$ are present in equal proportions (this is actually rather uncommon, but let's go with it for now). 

If you were to pick a sequence of 10nt completely at random, what is the chance that exactly 3 bases will be an $A$ (or any other homopolymer)?
```{r}
dbinom(3,10,.25)
```

What is the chance that you will find less than 5 $G$s (i.e. 4 or fewer)?
```{r}
pbinom(4,10,.25)
```

What is the chance that you will find between 2 and 4 $T$s?
$$ P(2 \le X \le 4) = \sum_{k=2}^4 {10 \choose k}(0.25)^k(0.75)^{10-k} = \sum_{k=0}^4 {10 \choose k}(0.25)^k(0.75)^{10-k} - \sum_{k=0}^1 {10 \choose k}(0.25)^k(0.75)^{10-k}$$

```{r}

# these are equivalent: a) sum up the probabilities for all values
sum(dbinom(2:4,10,.25))
# b) get the cumulative mass up to and including 4, then
#    subtract the cumulative up to 1 (to get 2 inclusive)
pbinom(4,10,.25) - pbinom(1,10,.25)
```

The quantile function `qbinom(p, size, prob)` returns the smallest value of $q$ such that $Pr(X \le q) \ge p$. The quantile is defined as the smallest value $x$ such that $F(x) \ge p$. For example, $F(x) \ge 0.75$ means that 75% of the distribution is less than $x$. Here, you are likely to find 3 or fewer of each kind of base in 75% of 10-mers picked at random from the genome:

```{r}
qbinom(.75,10,.25)
```

What is the maximum number of any single base you would expect to find in 99% of random 10-mers?

```{r}
qbinom(.99,10,.25)
```


[^1]: Disclaimer: I made this up, but based on a true story (see https://en.wikipedia.org/wiki/Peppered_moth_evolution).