---
title: 'Homework 5: Contingency Tests'
subtitle: "XDASI Fall 2021"
author: "Florence Nightingale [YOUR NAME HERE]"
date: "Due: 10/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE)
```


# Association between obesity and genetic polymorphisms

## Q0: Load BMI Data

This is a published data set containing measurements of Body Mass Index (BMI) and single-nucleotide polymorphisms (SNPs) based on genotyping of 79 individuals. The data file provided contains the BMI (numeric) and the presence or absence of 7 SNPs (a factor, encoded as Y or N) for each individual.

Our goal in this study is to determine whether any of the SNPs are associated with Obesity.

#### a) Load the `HW6_SNP_Data.tsv` file into a dataframe called `hw6.data`.

```{r}

```

#### b) Briefly describe the structure of this observational data set. 

What measurements have been taken, and what is the data type of those measurements?

```{r eval=FALSE}
# your answer here
```


## Q1: $\chi^2$ Test

#### a) Add an column for `Overweight` status

The column should have binary values (e.g. TRUE/FALSE or Y/N) according to whether the BMI for each individual is 25 or above, or under 25.


```{r}

```

#### How many individuals are overweight by this definition? How many are not overweight?

```{r}
table(hw6.data$Overweight)
```

#### b) Generate a contingency table showing Overweight and the presence of SNP1.
**Hint - you may choose to use the table() function**
```{r}
snp1.contingency.table <- table(hw6.data[, c("Overweight", "SNP1")])
snp1.contingency.table
```

#### c) Generate an expectation table showing the EXPECTED number of individuals when comparing Overweight to SNP1. Calculate this table using the formula discussed in class.
$$E_{A{\cap}B} = Pr_A * Pr_B * Total$$

```{r}

# The totals
obesity = table(hw6.data$Overweight)
snp1present=table(hw6.data$SNP1)

expected_snp1 = matrix(NA, nrow=2,ncol=2)
expected_snp1[1,1]=obesity[1]*snp1present[1]/sum(obesity)
expected_snp1[2,1]=obesity[2]*snp1present[1]/sum(obesity)
expected_snp1[1,2]=obesity[1]*snp1present[2]/sum(obesity)
expected_snp1[2,2]=obesity[2]*snp1present[2]/sum(obesity)
rownames(expected_snp1)=c("False","True")
colnames(expected_snp1)=c("N","Y")
```

```{r}
print(expected_snp1)
```
#### d) Calculate the $\chi^2$ statistic

```{r}
snp1_chisq_statistic = sum(((snp1.contingency.table - expected_snp1)**2)/expected_snp1)
snp1_chisq_statistic
```


#### e) Test your observed contingency table for Overweight vs SNP1 with the Chi-Squared test. Use the `pchisq()` function with the $\chi^2$ test calculated in the previous step.

```{r}
pchisq(snp1_chisq_statistic, 1, lower.tail = F)

```
#### f) Confirm your results using the `chisq.test()`
Be sure to turn the correction off since we did not perform the correction in the calculation above.

```{r}

chisq.test(snp1.contingency.table, correct=F)

```

#### g) What is the null hypothesis for this chi-squared test?


```{r eval=FALSE}
# your answer here
The two variables (Overweight & SNP1) are independent.
```


#### h) Interpret the result of your chi-square test between the Overweight trait and SNP1.

```{r eval=FALSE}
# your answer here
We fail to reject the null hypothesis; it appears that SNP1 is unlikely to have an effect on obesity.
```

## Q2: Relative Risk vs Odds Ratio

#### a) Calculate the **relative risk** for SNP1 and Obesity.

```{r}
#prob of having SNP1 in obese / prob of having snp1 in not obese

risk1=snp1.contingency.table[1,2]/sum(snp1.contingency.table[1,])
risk2=snp1.contingency.table[2,2]/sum(snp1.contingency.table[2,])

relrisk = risk2/risk1
relrisk
```

#### b) Confirm your results using the `riskratio` function in the `epitools` package.


```{r}

library("epitools")
riskratio(snp1.contingency.table,correction = F)
```

#### c) Calculate the **odds ratio** ratio for SNP1 and Obesity.

```{r}
# odds ratio is the ratio of the odds
# odds that a obese person has snp1 is
odds1= snp1.contingency.table[1,2] / snp1.contingency.table[1,1] 

odds2 =snp1.contingency.table[2,2]/ snp1.contingency.table[2,1]

odd_ratio = odds2/odds1

odd_ratio
```

#### d) Confirm your results using the `oddsratio` function in the `epitools` package.

Make sure you use the `wald` method in the `oddsratio()` function.

```{r}
install.packages("epitools")  # do this once
library(epitools)
oddsratio(snp1.contingency.table, method="wald")
```

## Q3: Fisher's exact test

#### a) Test the relationship between the Overweight trait and SNP1 with a Fisher's Exact Test

```{r}
snp1.fisher <- fisher.test(snp1.contingency.table)
print(snp1.fisher)
```

#### b) Describe the difference between the Fisher Exact Test and the Chi-Squared test

```{r eval=FALSE}
# your answer here

# Fisher's exact test is nonparametric and calculates p-values based on the hypergeometric distribution.

# The Chi-squared test is parametric and uses the normal approximation for a binomial distribution.
# The p-value for the test statistic, computed as sum((observed - expected)^2 / expected),
#   is found using a Chi-squared distribution with (r-1)(c-1) degrees of freedom.
```

#### c)What is the null hypothesis for this Fisher's Exact Test?

<!-- INSERT ANSWER HERE -->

The two variables are independent

#### d) Interpret the result of your Fisher's Exact test. Compare the p-values to the chi-squared test.

<!-- INSERT ANSWER HERE -->


#### e) Check the other SNPs to see which SNP is most associated with obesity.

You may use the Fisher's eExact tTest to check for this.


```{r}
for (i in 2:ncol(hw6.data)) {
  print(colnames(hw6.data)[i])
  snp.contingency.table <- table(hw6.data[, c("Overweight", colnames(hw6.data)[i])])
  snp.fisher <- fisher.test(snp.contingency.table)
  print(snp.fisher$p.value)
}


```

Are any of the SNPs significantly asociated with BMI?

```
<!-- INSERT YOUR ANSWER -->
```