---
title: 'XDAS 2020: In-class exercises: Probability'
author: "Kris Gunsalus"
date: "9/24/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE)
# remove "error = TRUE" to make knitr halt on errors
# (do this when you are ready to check if your code will knit without errors)
```

## Practice problems

### 1. Co-occurrence of infection: Venn diagams

Among women voluntarily tested for sexually transmitted diseases in one university (modified from TÃ¡bora et al. 2005):

+ 18% tested positive for human papilloma virus (HPV) only,
+ 2% tested positive for Chlamydia only, and
+ 3% tested positive for both HPV and Chlamydia. 

Use the following steps to calculate the probability that a woman from this population who gets tested would test positive for either HPV or Chlamydia.

a. Write the goal of the question using probability notation (e.g. P(something) or Pr[something]). 

<!-- Your answer here -->
*What is P(HPV OR Chlamydia)?*

b. Write the general addition rule, as applied to this example.

<!-- Your answer here -->
*P(HPV OR Chlamydia) = P(HPV) + P(Chlamydia) - P(HPV AND Chlamydia)*

c. First, calculate the total frequency of HPV and of Chlamydia in the population.
```{r}
# probabilities
p.hpv = 0.18
p.chl = 0.02
p.hpv.chl = 0.03

p.hpv.tot = p.hpv + p.hpv.chl
p.chl.tot = p.chl + p.hpv.chl
p.hpv.tot
p.chl.tot
```

d. Calculate the probability that a randomly sampled woman would test positive for at least one of these, using the total probabilities and the joint probability.

```{r cars}

# P(HPV OR Chlamydia)
p.hpv.tot + p.chl.tot - p.hpv.chl

# this is the same as adding up the original numbers
p.hpv + p.chl + p.hpv.chl
```

e. Is the occurrence of these infections independent? Explain.
```{r}
# Independence: P(HPV AND Chlamidia) = P(HPV) * P(Chlamydia)
indep.test = p.hpv.tot * p.chl.tot
indep.test
indep.test == p.hpv.chl
```


### 2. Allele frequencies: Addition and multiplication rules

Many gene loci have a major allele and a number of minor alleles in the population. Let's say there are 5 alleles for a particular locus, that individuals mate randomly with respect to the genotype at this locus, and that allele frequences for A1-A5 are 0.82, 0.06, 0.05, 0.04, and 0.03 respectively.

For each of the questions below, write out the probability equations (e.g. Pr[Ai] or P(Ai) for each term) and then compute the answers numerically.

a. What is the probability that any single allele chosen at random from the population is either A4 or A5? Write out the probability statement as a comment below and then computer the answer.
```{r}
# Allele frequencies
A1 = 0.82
A2 = 0.06
A3 = 0.05
A4 = 0.04
A5 = 0.03

# Probability of either A4 or A5 = ?

# P(A4 OR A5) = P(A4) + P(A5)
A4 + A5
```

b. What is the probability that an individual carries two A2 alleles?
```{r}
# P(A2 AND A2) = P(A2)*P(A2)
A2*A2
```

c. What is the probability that someone does NOT carry two A2 alleles?
```{r}
#P(not (A2 AND A2) = 1 - P(A2 AND A2)
1 - A2^2
```

d. What is the probability that someone is heterozygous for A1 and A3?
```{r}
# P(A1 AND A3) = 2 * P(A1)*P(A3)
2*A1*A3
```

e. What is the probability that *neither* of two random individuals in the population would carry two A1 alleles?
```{r}
# P(2 individuals are NOT A1;A1) = (1 - P(A1;A1))^2
(1-A1*A1)^2
```

f. What is the probability that 3 random individuals carry no A4 or A5 alleles at all? (Remember that each individual carries two alleles.) 
```{r}
# P(A4 or A5) = P(A4) + P(A5)
# P(NOT (A4 or A5)) = 1 - P(A4 or A5)
# P(3 are not A4 or A5) = ( 1 - P(A4 or A5) )^3
p4or5 = A4 + A5
pNOT4or5 = 1 - p4or5
pNOT4or5^6 # 3 invididuals * 2 alleles each
```

### 3. Nucleic acids: Sampling, permutations and combinations

a. Restriction enzymes recognize specific sequences in DNA and cut the DNA within or near those sites. How many possible restriction sites of length 6 are there?
```{r}
4^6

# There are 4 possibilities for each position, and they are drawn independently, so there are 4*4*4*4*4*4 = 4^6 possibilities.
```

b. At what frequency would you expect to find a binding site for a particular restriction enzyme that recognizes a 6 bp sequence? _Note: Restriction enzymes usually recognize a palindromic sequence, so you don't need to worry about looking on both strands._
```{r}
1/4^6

# Since there are 4096 possible 6-mers, you would expect to find a specific one around 1/4096th of the time. This is 2.4 sites per 10,000 bases, or one site every 4096 bases.
```

c. What is the average length of a fragment produced by a 6-cutter?
```{r}
4^6

# One cut every 4096 bases on average translates to an expected fragment size of 4096.
```

d. The EcoR1 enzyme cuts the sequence "GAATTC". How many different sequences could be made out of this set of nucleotides (assume that you treat each A and T as distinct individuals)?
```{r}
# nPerm = n!
factorial(6) # 720
```

d. You are interested in synthesizing a bunch of random oligonucleotides for a SELEX experiment. How many different oligos of length 22 could you synthesize?
```{r}
4^22 # 1.76e+13
```


e. How many ways could you make a sequence of 6 nt by grabbing them at random from a bag of 12 nucleotides (assuming each base is treated as a distinct unique entity, and you can only pick each nt once)? This is an example of "random sampling without replacement".
```{r}
# nPerm = n!/(n-k)!
factorial(12)/factorial(12-6) # 665280
```

f. How many combinations of 6nt are there in a set of 12 random nucleotides? (That is, pick any group of 6 nt, where each one is treated as unique, and you don't care about the sequence in which you pick them.)
```{r}
# nComb = nPerm/k! = n!/k!(n-k)!
factorial(12)/( factorial(6) * factorial(12-6) ) # 924
```

g. How are your answers in parts e and f related?

<!-- Your answer here -->
*The number of combinations is smaller than the number of permutations by a factor of k! since all the permutations of the k-mer sequence get collapsed into one combination.*
