---
title: "XDAS19_Final_Exam"
author: "Justin Blau"
output:
  html_document:
    toc: yes
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE, fig.align = 'left')

# HINT: Package loading is not provided for you
```

---

#### PCA and Clustering are useful in grouping together genes and samples based on the the differences in the values and expression patterns.

The distance between datapoints can be calculated in many ways, and the method of choice reflects the question one wants to ask. In the case of gene expression, one often looks for genes that are covariant across the experiments observed. 

A popular method of calculating covariance-based distance is the ``Pearson method``. In some cases we will be looking at the actual geometric difference. When we center (subract from the mean) and scale (divide by standard deviation), we are essentially making Euclidean (geometric) distance equivalent to Pearson distance.

#### In this example, we will look at a publicly available corn microarray dataset.

Yang XS, Wu J, Ziegler TE, Yang X et al. Gene expression biomarkers provide sensitive indicators of in planta nitrogen status in maize. Plant Physiol 2011 Dec;157(4):1841-52.

The object of this project is to identify biomarkers that can be used to determine whether a plant is using Nitrogen efficiently. Also it is known that plants are more efficient in nutrient uptake at different times of the day, so the sceintist designed an experiment where the samples were treated at different times. Finally another factor in the experimental design is genotype. The scientist decided to look at the different genotypes to see if one was better at nitrogen use efficiency.

**In total there are 90 samples:**

+ **Genotype:** 4 different lines labaled 1-4

+ **Nitrogen:** 3 different treatments : 2mM, 20mM, and grown 2mM and then treated with 20mM

+ **Time:** Day1-10am, Day1-11pm, and Day2-10am.

**Each combination has 3 replicates.**


# Q0: Loading the data (10 points)

#### GEO is the Gene Expression Omnibus. It is hosted by the National Center for Biotechnology Information (NCBI). It is a database of gene expression data sets. 


### A) Dowload the GEO dataset

#### Install the GEOquery package and retrieve the GSE32361 data set from GEO using the `getGEO` function from GEOquery. It is expression data from corn plants.

```{r}
# BiocManager::install("GEOquery")
gse<-getGEO(___)
```


#### `getGEO` returns a *list* of ExpressionSet *objects*. 

The **ExpressionSet** class is a container with *arguments* called *slots* that store complex information about expression set data. ExpressionSet is an extension of a more generic class called **eSet**, which stores data about high-throughput experiments. Documentation for ExpressionSet may be found here: https://www.rdocumentation.org/packages/Biobase/versions/2.32.0/topics/ExpressionSet

Each ExpressionSet object stores information such as the **expression values** (`assayData`, a *matrix*) and **metadata** about the experiment set (`phenoData`, a *data.frame*). Metadata may include additional information about experimental conditions, protocols, publication and contact data, etc. 

ExpressionSet also comes with a set of class-specific _**methods**_ to retrieve various items of interest (see R documentation). Fortunately for us we will be using these helper functions (`exprs` and `pData` below) that will allow us to retrieve the data without worrying to much about how the data is stored.


#### The `getGEO()` command you just used gave a *list* with just one ExpressionSet *object* that contains seven *slots*.

Explore the structure of the object you retrieved to get a feel for how the data are organized (do this on your own in the console).


### B) Extract the *expression data* and *metadata*

#### Perform the following:

+ Use the `exprs` method on the ExpressionSet object in the list to retrieve the *assay data* (as a matrix), and then
+ convert the expression data into a data.frame (you will *not* need to keep the intermediate matrix for later use).
+ Use the `pData` method to retrieve the metadata for the *samples* in a data.frame.

```{r}
expression.data <- ___
expression.metadata <- ___
```


### C) Extract the three important *metadata* columns

The three columns we care about contain the plant **genotypes**, the **growth nutrient**, and the **sampling time**. Identify what these are called and then select them to create a four-column data frame. This dataframe must have the GEO accession number (GSM ID) and each of the three important metadata columns.

```{r}

```


### D) Data validation

#### Confirm the following statements using R. 

**Answers that are not the result of functional tests written in R will receive no credit.**

+ The number of samples in the expression data matches the number of samples in the metadata. 

```{r}

```

+ Each expression data sample ID has a corresponding record in the metadata.

```{r}

```

+ Each expression data sample ID is unique.

```{r}

```


# Q1: Distance & Clustering (20 points)

### A) Create a cluster dendrogram

+ Use `hclust` to cluster the samples by *Euclidean* distance using the *average linkage* method. 
+ Plot a dendrogram labeled with the **sample IDs** (use the `ggdendro` package, which extends `ggplot2`).

```{r}

```


#### Briefly explain what a dendrogram shows.

<!-- INSERT ANSWER HERE -->


##### There is definitely some grouping for the samples. Let's take a closer look at the metadata of the samples to see if we can understand what is going on.

The columns are now grouped based on similarity of expression. Below we will try to see which aspects of the experimental design have the most influence on the clustering outcome.

##### For this exercise, you will need to figure out how to color the labels on the dengrogram using different sets of *factors*. 

**HINTS:**

+ First, explore the `hclust` function on your own to get a feel for how the result object is structured (not part of the answer). 
  - Note that you can get the *names* and *order* of the samples in the dendrogram directly from the `hclust` object.
+ You can easily color labels on `ggplot` objects using a vector of color names.
  - For example, if you have 10 labels on the X axis, you need to create a vector with colors for all 10 labels, e.g. `color.vector <- c("black", "blue", "red", "green", "black", "blue", "red", "green", "black", "black")`. 
  - You can then color the labels by adding to the ggplot: `theme(axis.text.x = element_text(color=color.vector))`.


### B) Explore patterns in the data using the dendrogram.

Below we will replot the above dendrogram with the *sample ID* labels colored based on different *factor* groupings.

#### First, replot the above dendrogram with sample ID labels colored by *genotype*.

```{r}

```

#### Replot the above dendrogram with sample ID labels colored by *growth nutrient*.

```{r}

```

#### Replot the above dendrogram with sample ID labels colored by *sampling time*.

```{r}

```


### C) Interpret patterns in the data using the dendrograms

#### Based on these `hclust` plots, of the three metadata features, what is the most important for gene expression? What is second most important? What is least important? How can you tell?

<!-- INSERT ANSWER HERE -->


# Q2: Gene Expression (45 points)

### A) Remove genes with low expression

Calculate the sum of the expression values for each gene. Keep only those genes that have expression sum greater than the median expression sum of all genes. As discussed in class, genes with low expression are unreliable.

```{r}

```


### B) ANOVA

Since we have three different factors, we should perform a 3-way ANOVA analysis to see which combination of factors best explains the expression patterns for each gene.

In our case we will test to see if each gene has a 3-way interaction and record the p-value of the model reported.

#### Perform 3-way ANOVA for highly expressed genes.

```{r}
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
# The following function will return the smallest p-value associated with a non-intercept model term
# This is different from the model F-statistic which does incorporate the intercept term
# There are better ways to do this; use this way because it's relatively fast and kinda ok
# Arguments:
#   1) a vector of expression values
#   2) a dataframe with the three factors as three columns
msk3wayAnova <- function(expvalue, expfactors) {
  templm<- lm(as.numeric(expvalue) ~ as.factor(expfactors[,1]) +
              as.factor(expfactors[,2]) +
              as.factor(expfactors[,3]))
  return(min(summary(templm)$coef[-1,4]))
}
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #

# Apply the above function to each row of the expression values.
anovaPvalues <- ___
```


### C) Extract significantly varying genes

We are most interested in genes that show significant differences in gene expression across all the treatments.
 
+ Identify the 1000 genes which have an ANOVA term that is most statistically significant (smallest p-value).
+ Create a data frame with only these 1000 genes (make sure to include all sample data for these genes; this data frame should be 1000 x 90).

```{r}

```


#### Plot a heatmap of these treatment-linked genes. Use the z-score of the gene expression changes as the fill.

**HINT:** The plot should contain 1000 rows by 90 columns. It is not helpful to plot gene names or to plot sample names. Remove the labels from this plot for now.

```{r}

```


### D) Explore patterns in the data using the heatmap

Below we will reorder the above plot in different ways to see how different treatment variables affect the strongest gene expression changes.

**HINTS:** 

+ To change the order of a discrete axis for a ggplot object:
  - Create a vector with the order you'd like, e.g. `gene.vec <- c("GeneA, "GeneC", "GeneB")`
  - Set: `scale_y_discrete(limits=gene.vec)`
+ You do not necessarily need to make a new `ggplot` object for each question. Instead, you can just change the display order for the plot object you created above and reassign the new output to the same object.

#### Reorder the *genes* in the heatmap based on heirarchial clustering, using Euclidean distance as the metric. Plot the new heatmap. 

**HINTS:** 

+ Use the standard normal expression values (the *z-scores* you calculated above) to cluster the *genes* baed on their overall similarity across all the conditions. 
+ For this question, the ordered vector you need to use to change the display can be obtained from the results of the `hclust` function.

```{r}

```

#### Reorder the biological *samples* in the heatmap based on the *genotypes*. Print the new heatmap.

```{r}

```

#### Reorder the biological *samples* in the heatmap based on the *growth nutrients*. Print the new heatmap.

```{r}

```

#### Reorder the biological *samples* in the heatmap based on the *sample time*. Print the new heatmap.

```{r}

```


### E) Interpret patterns using the heatmaps

#### How can you tell from a heatmap which genes are varying because of one of the three experimental factors (i.e. genotype, growth nutrients, or sample time)? Are there any that are varying because of growth nutrients? How can you tell? Are there any that are varying because of sample time? How can you tell?

<!-- INSERT ANSWER HERE -->


### F) Create discrete clusters

#### Using the heatmaps in the previous section, how many distinct clusters of genes do you think are there? Explain your answer. Remember that every row is a gene and every column is a different sample.

<!-- INSERT ANSWER HERE -->

#### Use the `cutree` function to create the number of groups you guessed above.

```{r}

```

#### Create a heatmap with only the genes in the largest group. Order the x-axis by `sampletime`.

```{r}

```

#### Describe the expression pattern for this group.

<!-- INSERT ANSWER HERE -->


### G) Use Silhouette plots to evaluate cluster quality

Load the `cluster` package so you can use the `silhouette` command.

#### Use a `for` loop to generate 3, 4, or 5 clusters (using `cutree`) and then generate a silhouette plot for each alternative grouping.

```{r}
# HINT: when plotting the silhouette use border=NA to make sure it shows properly.
```

#### Compare the results to your original guess for the optimal number of groups. Which number of groups better fits the data? Explain.

<!-- INSERT ANSWER HERE -->


# Q3: PCA (25 points)

### A) Principal Components

#### Calculate the principal components for the 1000 treatment-linked genes from Q2.

```{r}

```


### B) Plot PC1 and PC2 

#### Use both *growth nutrient* and *sampling time* to label the data on the plot. 

**NOTE:** You should choose whatever visual cues you think would be most helpful for illustrating the differences between treatments. This is an important choice when getting ready to publish your data.

```{r}

```


### C) Interpret the data using the PCA plot

#### What does this PCA plot tell you about this data? Of the three known experimental factors, what is the most important effect on gene expression? What is the next most important effect on gene expression? How can you tell?

<!-- INSERT ANSWER HERE -->


### D) Is this result consistent with the results from Q1 and Q2?

<!-- INSERT ANSWER HERE -->
