---
title: 'Homework 9: Exponential and Negative Binomial Distributions'
author: "Mari Shiratori"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: default
  html_document: default
subtitle: XDASI Fall 2021
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(dplyr)
library(ggpubr)
```

## Q1: FRAP Experiment

Fluorescence Recovery After Photobleaching (FRAP) allows the measurement of how fast molecules can move around in cells. When a fluorophore located in a particular region of a cell is bleached out with a laser, the rate at which fluorescence reappears provides a measure of how fast the molecule can diffuse back into the bleached spot under different conditions.

Germ granules play a role in specifying the germline and in maintaining germline immortality in animals. In *Drosophila*, recruitment of the Vasa RNA helicase to germ plasm is needed for proper germline development. 

An experiment to determine whether a *C. elegans* protein, MIP-1, helps anchor GLH-1 (a Vasa RNA helicase homolog) in germ granules of the *C. elegans* early embryo was performed. The experiment used FRAP to compare recovery rates of GLH-1::GFP within germ granules in WT and *mip-1* mutant embryos.[^1]

### a. Load the dataset and examine it

The data file "glh_frap.csv" contains FRAP data for GLH-1::GFP in WT or *mip-1* mutant embryos. Load the dataset and take a look at it.

```{r}
frap <- read.csv("./glh_frap.csv")
str(frap)
head(frap)
```


### b. Make a quick line plot

To check how the data look, first make a summary table of the mean `Intensity` for each timepoint across the 8 replicates for each strain. Then, use the summary table to plot `Intensity` vs. `Time` with `ggplot2` (color the lines by `strain`).

```{r}
# summary table
frap_sum_table <- frap %>%
  group_by(Time, strain) %>%
  summarize(Intensity_mean = mean(Intensity))

#str(frap_sum_table)
head(frap_sum_table)

# plot
ggplot(data = frap_sum_table, mapping = aes(x = Time, y = Intensity_mean, color = strain)) +
  geom_line()

```


### c. Normalize the data and replot

This looks like a good start, but we need to normalize the data in order to be able to compare the recovery rates. To do this we will rescale the data so that the minimum value for both strains is 0, and the intensities are relative to the starting value of 1.

***(i) Rescale the data***

First, create a very small function to normalize the Intensity using the formula `(x-min(x))/(1-min(x))`. The function should take a single numerical argument and return the result of the formula.

```{r}
# normalization function
norm_fun <- function(x) {
  return((x - min(x))/(1 - min(x)))
}

norm_fun(frap_sum_table$Intensity_mean)
```

Next, add a new column to the original data table containing the normalized Intensity. To get this right, you will need to do the normalization separately for each set of replicate measurements for each strain. 

```{r}
# add normalized intensity column
norm_table <- frap %>%
  group_by(rep, strain) %>%
  mutate(Norm_Intensity = norm_fun(Intensity))

# View(frap.norm.summary)
head(norm_table)

```

Check your work by examining the values in the `norm.Intensity` column. These should be `1` at time `-1.0`, and `0` at time `0.0` for each of the 8 replicates from both strains. 


***ii) Replot using the normalized data***

Make a new summary table and make the same plot as in part (b) using the normalized data.

```{r}
# summary table
frap_norm_table <- norm_table %>%
  group_by(Time, strain) %>%
  summarize(Norm_Int_mean = mean(Norm_Intensity))

head(frap_norm_table)

# plot
ggplot(data = frap_norm_table, mapping = aes(x = Time, y = Norm_Int_mean, color = strain)) +
  geom_line()
```

You will notice that the maximum intensity upon recovery in the mutant strain seems to be slightly above 1. Don't worry about this for now.


### d. SD, SEM, and 95%CI

***i) First, add 3 new columns containing the SD, SEM, and number of repeats per condition to the summary table.***

```{r}
# summary table with columns for mean, SD, SEM, and number of replicates
frap_norm_table <- norm_table %>%
  group_by(Time, strain) %>%
  summarize(Norm_Int_mean = mean(Norm_Intensity),
            Norm_Int_sd = sd(Norm_Intensity),
            Norm_Int_sem = sd(Norm_Intensity)/sqrt(max(rep)),
            replicates = max(rep))

frap_norm_table
head(norm_table)
```

***ii) Add columns containing the upper and lower bounds for the SD, SEM, and 95%CI to the summary table.***

Note: since these are small samples, you will want to use the $t$-distribution to compute the 95%CI for these.

```{r}
# SD.upper and SD.lower
# SE.upper and SE.lower
# CI.upper and CI.lower
# I put all of them together because it was easier for me to to do this rather than adding the mutations separately. I hope that's okay.
frap_norm_table_new <- frap_norm_table %>%
  mutate(sd.upper = Norm_Int_mean + Norm_Int_sd,
         sd.lower = Norm_Int_mean - Norm_Int_sd,
         se.upper = Norm_Int_mean + Norm_Int_sem,
         se.lower = Norm_Int_mean - Norm_Int_sem,
         ci.upper = Norm_Int_mean + qt(0.975, replicates - 1) * Norm_Int_sem,
         ci.lower = Norm_Int_mean - qt(0.975, replicates - 1) * Norm_Int_sem)

head(frap_norm_table_new)


```

***iii) Make 3 ribbon plots showing the normalized recovery curves plus the SD, SEM, and 95%CI.***

```{r}
# wes anderson palettes loaded to make the figs pretty.
#install.packages("wesanderson")
require(wesanderson)

# SD plot
sd.plot = ggplot(data = frap_norm_table_new, mapping = aes(x = Time, y = Norm_Int_mean, col = strain)) +
  geom_ribbon(aes(ymin = sd.lower, ymax = sd.upper, fill = strain), alpha = 0.4) +
  geom_line() +
  scale_color_manual(values = wes_palette("Darjeeling2", n = 2)) +
  scale_fill_manual(values = wes_palette("Darjeeling2", n = 2)) +
  theme_classic() +
  ggtitle("FRAP with SD")

  sd.plot

# SEM plot
sem.plot = ggplot(data = frap_norm_table_new, mapping = aes(x = Time, y = Norm_Int_mean, col = strain)) +
  geom_ribbon(aes(ymin = se.lower, ymax = se.upper, fill = strain), alpha = 0.4) +
  geom_line() +
  scale_color_manual(values = wes_palette("Darjeeling2", n = 2)) +
  scale_fill_manual(values = wes_palette("Darjeeling2", n = 2)) +
  theme_classic() +
  ggtitle("FRAP with SEM")

# 95%CI plot
ci.plot = ggplot(data = frap_norm_table_new, mapping = aes(x = Time, y = Norm_Int_mean, col = strain)) +
  geom_ribbon(aes(ymin = ci.lower, ymax = ci.upper, fill = strain), alpha = 0.4) +
  geom_line() +
  scale_color_manual(values = wes_palette("Darjeeling2", n = 2)) +
  scale_fill_manual(values = wes_palette("Darjeeling2", n = 2)) +
  theme_classic() +
  ggtitle("FRAP with 95%CI")

ggarrange(sd.plot, sem.plot, ci.plot, nrow=1, common.legend=TRUE)
```

***iv) Which graph would you prefer to show in your paper, and why?***

```{r eval=FALSE}
# I'd prefer to show the SEM ribbon graph if I were to publish this in a paper because the ribbons don't overlap at all while they do in the other two graphs.
```


### e. Test for significant differences in recovery rates

These curves look pretty different. One way to quantify this is to calculate the recovery rates and compare them using a statistical test.

The FRAP curves follow the equation $I(t) = A(1-e^{-kt})$, where I(t) is the measured intensity as a function of time t and A is the maximum recovered intensity. The file "rates_glh.csv" contains the recovery rates computed by a FRAP software package.

***i) Boxplot + strip chart***

Load this dataset and make a boxplot of the rates, grouped by strain. Overlay the data actual points onto these. (What happens if you overlay the boxes on top of the points instead?)

```{r}
# load FRAP rates
rates <- read.csv("./rates_glh.csv")
head(rates)

# make the plot
ggplot(data = rates, mapping = aes(x = Treatment, y = Rate, color = Treatment)) +
  geom_boxplot(aes(color = Treatment)) +
  scale_color_manual(values = wes_palette("Darjeeling2", n = 2)) +
  geom_point() 

```


***ii) Use an appropriate statistical test to compare the differences in the rates between the WT and mutant strains.***

```{r}
# get WT and mutant rates separately
WT_rate <- rates$Rate[which(rates$Treatment == "WT")]
mutant_rate <- rates$Rate[which(rates$Treatment == "mip1uae1")]

# test for significance
t.test(x = mutant_rate, y = WT_rate, alternative = "g")

```


***iii) What can you conclude about the effect of the MIP-1 protein on GLH-1 within germ granules?***

```{r eval=FALSE}
# MIP-1 does indeed help anchor GLH-1 in germ granules of the C. elegans early embryo. This is evident because the rates as calculated above are significantly different. 
```


## Q2: Stress Response

Expression of many genes is controlled at the translational level. One example is the stress response gene ATF4. In unstressed conditions, ribosomes bind to an ATF4 transcript at the 5' end, and then translate an upstream Open Reading Frame (uORF), which is a short sequence that does not produce functional protein. 

At the stop codon of this uORF, in unstressed conditions, the ribosome dissociates from the transcript **98% of the time**. If the ribosome does not dissociate, it re-initiates translation and produces a functional ATF4 protein.

### a. Probability mass function for 1 "success"

Recall that the negative binomial PMF describes the probability of $r$ 'successes' after $x$ 'failures.' The random variable for this distribution is $X_r$, the number of unsuccessful trials before getting a desired number of successes.

***i) What is the probability of 'success' in this experiment?***

Hint: To set this problem up correctly for the NB distribution, here we define "failures" to be the number of correct dissociations, and "success" to be the production of a functional ATF4 protein.

```{r eval=FALSE}
# The probability of success is 0.02 because the ribosome dissociates from the transcript 98% of the time.
```


***ii) Density Plot***

Plot the PMF for the ***number of ribosomes*** that correctly dissociate from the ATF4 transcript before a ribosome erroneously produces ***one functional ATF4 protein molecule***. 

**Note: For the plots in this question, we use LINE plots (even though the NB is a discrete distribution) to make it easy to visualize the shape of the distribution.**

Recall that the `binom` family of functions takes three parameters:

+ x = The total number of failures before r successes
+ size = r (the target for the number of successes)
+ prob = probability of SUCCESS

```{r}
# Create the "ideal" NB probability mass function over the range x=0:200.
x = c(0:200)
p = 0.02

pmf <- dnbinom(x, 1, p)
head(pmf)


# Create a data frame with two columns: the range vector, and
# the corresponding PMF values.
range <- seq(0,200)

pmf_df <- as.data.frame(cbind(range, pmf))
str(pmf_df)

# Make a density plot (the range goes on the x-axis, the density goes on the y-axis).
ggplot(data = pmf_df, mapping = aes(x = range, y = pmf)) +
  geom_line(color = "rosybrown2")

```

***iii) This particular NB example may also be modeled by another distribution we talked about in class. What is it, and why? Explain.***

```{r eval=FALSE}
# Geometric because similarly to the negative binomial, they are both related to the binomial distribution where there are independent trials with just 2 possible outcomes. 
```


### b. Activation rate for stress response

Each ATF4 transcript must produce **10 functional ATF4 protein molecules** to activate the cellular stress response program. Here we will examine the _**rate of activation**_ of the cellular stress response program. 

***i) PMF plot: ribosome binding events upon activation of the stress response***

First, plot the PMF for the **number of ribosome binding events** that will occur before **10 functional ATF4 protein molecules** are made.

```{r}
# Create the probability mass function over the range 0:1000 (ribosome bindings)
# for r=10 "successes"
x2 = c(0:1000)
r2 = 10
p2 = 0.02


pmf_1000 <- dnbinom(x2, r2, p2)

# Create a data.frame with the data
range_1000 <- c(0:1000)
pmf_2_df <- as.data.frame(cbind(range_1000, pmf_1000))


# Make a density plot 
ggplot(data = pmf_2_df, mapping = aes(x = range_1000, y = pmf_1000)) +
  geom_line(color = "rosybrown2")

```

***ii) PMF plot: time to activation of the stress response***

Ribosomes bind to the transcript at a rate of **1 every 10 seconds**. Plot the PMF for the expected number of **MINUTES** it will take to activate the cellular stress response program.

First, add a Time column to your data frame containing the corresponding number of minutes across the range of ribosome bindings. Then, plot the PMF for the time to activation.

```{r}
# Add a Time column to this data frame giving the number of minutes to activation.
time_x = c(0:1000)*10/60

#time_x

pmf_2_df_time <- cbind(pmf_2_df, time_x)

head(pmf_2_df_time)


# Make a density plot of the TIME to activate the response
ggplot(data = pmf_2_df_time, mapping = aes(x = time_x, y = pmf_1000)) +
  geom_line(color = "rosybrown2")
```


### c. Post-translational regulation of ATF4

ATF4 is also regulated post-translationally; it is degraded with a half-life of 20 minutes. 

Based on the graph above, would you expect that the cellular stress response program is likely to activate by chance in unstressed conditions, due to leakage through the regulatory mechanisms? If the protein were very stable, how would your answer change?

Answer qualitatively in ~2-3 sentences (you do not need to create any additional mathematical models to answer this).

```{r eval=FALSE}
# If the protein were very stable, it would likely not have leakage to activate by chance in unstressed conditions. Unstressed even without a super stable protein, it's also unlikely for there to be leakage.

```


### d. Regulation by signaling

You add 4 micrograms per milliliter of tunicamycin to the culture, causing signalling pathways to phosphorylate eIF2alpha.

The ribosome now dissociates from the uORF 60% of the time and re-initiates to create a functional ATF4 protein 40% of the time. (Note that it still takes 10 productive ATF protein molecules to activate the stress response, but the probability of re-initiation is now much higher.)

On the same plot, draw the PMF for the **TIME** it takes to activate the stress response program in the presence of tunicamycin and in unstressed conditions.

```{r}
# First, add a 'Condition' column to the "activated" data frame and set it to
# "Unstressed"
condition_df <- pmf_2_df_time %>%
  mutate(Condition = rep("unstressed", nrow(pmf_2_df_time)))
head(condition_df)

# Create the PMF for the stressed condition over the range 0:1000 ribosome bindings
pmf_1000 <- dnbinom(c(0:1000), 10, 0.4)
head(pmf_1000)

# Create a data.frame with the data for the stressed condition.
# Add columns for Ribosome_Bindings, Time, and Condition.
stressed_df <- data.frame(range_1000 = c(0:1000), 
                             pmf_1000 = pmf_1000, 
                             time_x = time_x, 
                             Condition = rep("stressed", length(pmf_1000)))

# Row bind the unstressed and stressed data to create a full data set.
full_data <- rbind(condition_df, stressed_df)

# Make a combined density plot with 'Time' as the independent variable, and
# map the color to Condition.
ggplot(data = full_data, mapping = aes(x = time_x, y = pmf_1000, color = Condition)) +
  geom_line()

```


### e. Do you expect that the cell will activate the cellular stress response program now? Why or why not? Answer in no more than one paragraph (5 sentences).

```{r eval=FALSE}
# When there is stress, there is a high chance of leakage as we can see from the plot above where there's a huge short peak. There is a much lower 
```


[^1]: Cipriani et al., Novel LOTUS-domain proteins are organizational hubs that recruit *C. elegans* Vasa to germ granules. *eLife*, 2021 | doi:10.7554/eLife.60833