---
title: "Mouse genes"
author: "Kris Gunsalus"
date: "9/12/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


This dataset contains protein coding genes from the mouse genome along with their chromosomal coordinates and GC content. 

*Note: Non-standard chromosomes and ~1000 genes 200kb and longer have been removed to make the data easier to work with*


## 0. Load required libraries (ggplot and dplyr)
```{r}
library(ggplot2)
# install.packages("dplyr") # only do this once
library(dplyr)
```


## 1.Import the "mouse_genes.csv" file into an object called 'mgenes' and take a look at it.
```{r}
mgenes = read.csv("mouse_genes.csv", stringsAsFactors = FALSE)
str(mgenes)
```


## 2. Add a column called 'Gene_Length_kb' containing gene lengths. To make it easier to read, convert the lengths to kilobases, and round to the nearest integer. Take a look at the resulting data frame.

```{r}
mgenes$Gene_Length_kb = ( ( mgenes$End - mgenes$Start ) + 1 ) / 1000
mgenes$Gene_Length_kb = round(abs(mgenes$Gene_Length_kb), digits = 2)
head(mgenes)
```


## 3. Use ggplot to make a histogram of gene lengths.

+ make the bars a lighter color, add a black border to them, and add some transparency
+ try different bin widths (something between 1 and 10 looks good to me)
+ align the bars to the right of x=0 using `boundary = 0, closed = "left"`
  + this sets the left-most edge of the bars at x=0, and 
  + includes the left boundary in the bin (default is "right")
+ add axis labels

After you are happy with the graph, store it in an object called "mgenes_hist" and take a look at it.

```{r}
# original binning
ggplot(data = mgenes, aes(x=Gene_Length_kb)) +
  geom_histogram(col="black",fill="thistle", alpha=0.6,
                 binwidth = 6)

# shifted binning
mgenes_hist = ggplot(data = mgenes, aes(x=Gene_Length_kb)) +
  geom_histogram(col="black",fill="thistle", alpha=0.6,
                 binwidth = 6,
                 boundary = 0, closed="left") +
  labs(x="Mouse genes (kb)", y="Frequency")
mgenes_hist
```


## 4. Add some vertical lines to the  "genes_hist" graph using `geom_vline()`:

+ a dashed red line showing the the mean length (add color plus `linetype="dashed"`)
+ a solid red line showing the median length

Now take a look at the graph again. Do these data look skewed?

```{r}
mgenes_hist = mgenes_hist +
  geom_vline(aes(xintercept = mean(Gene_Length_kb)),
             color="red", linetype="dashed") +
  geom_vline(aes(xintercept = median(Gene_Length_kb)),
             color="red")
mgenes_hist
```


## 5. Add text labels to the lines calling them "mean" and "median".

```{r}
mgenes_hist = mgenes_hist +
  
  geom_text(aes(x= median(Gene_Length_kb) , # add this
                y=3000, label="Median"),
            color="red", angle=90, vjust = 1.5) +
  
  geom_text(aes(x= mean(Gene_Length_kb), # add this
                y=3000, label="Mean"),
            color="red", angle=90, vjust = 1.5)
mgenes_hist
mean(mgenes$Gene_Length_kb)
median(mgenes$Gene_Length_kb)
```


## 6. Use the `summary()` and `IQR()` function to find the IQR of gene lengths, and make a boxplot of gene lengths.

```{r}
summary(mgenes$Gene_Length_kb)
IQR(mgenes$Gene_Length_kb)

# first quartile
summary(mgenes$Gene_Length_kb)[2]
summary(mgenes$Gene_Length_kb)["1st Qu."]

boxplot(mgenes$Gene_Length_kb)
```


## 7. Remove the outliers and recalculate the mean and the median of gene lengths again.

+ Use `summary()` to get a vector of stats for the distribution and call it something short for convenience.
+ Then compute the min and max values
+ Use these to make a small dataframe that contains only the data between these values
+ Finally, compute the mean and median

```{r}
# set x to the `summary()` of mgenes$Gene_Length_kb for convenience below
# x is now a vector of stats for the distribution
x = summary(mgenes$Gene_Length_kb)

# calculate the highest and lowest values that are not outliers
# hint: where do the whiskers typically end on a boxplot?
min_len = x["1st Qu."] - 1.5*IQR(mgenes$Gene_Length_kb) # just the min val here
max_len = x[5] + 1.5*IQR(mgenes$Gene_Length_kb)
c(min_len,max_len)

# get trimmed data using base R
trimmed = mgenes[  which( mgenes$Gene_Length_kb >= min_len &
                          mgenes$Gene_Length_kb <= max_len   ),
                   "Gene_Length_kb" ]
head(trimmed)
tail(trimmed)

# OR, get trimmed data use dplyr
#library(dplyr)
trimmed = mgenes %>%                        # data frame
  filter(Gene_Length_kb >= min_len &
         Gene_Length_kb <= max_len) %>%     # filter rows
  select(Gene_Length_kb)                    # select columns
  
# mean and median of trimmed data
mean(mgenes$Gene_Length_kb)
median(mgenes$Gene_Length_kb)
mean(trimmed)
median(trimmed)
```


## 8. Mean and median of trimmed data

Add the same kind of mean and median vlines for the data without outliers to "mgenes_hist", using a reddish color.

+ Note that here, we are overlaying a different dataset onto the original plot
+ We do this by specifying the second dataset within the `geom` layer, and adding the `aes` mapping to that new dataset.

```{r}
mgenes_hist +
  geom_histogram(data = trimmed, aes(x=Gene_Length_kb),
                 col="black",fill="lightblue", alpha=0.4,
                 binwidth = 6) +
  
  geom_vline(data = trimmed, aes(xintercept = mean(Gene_Length_kb)),
             color="blue", linetype="dashed") +
  geom_vline(data = trimmed, aes(xintercept = median(Gene_Length_kb)),
             color="blue")


```


## 9. Variation in gene lengths by chromosome

Look at the variation in gene lengths among chromosomes by generating a ggplot containing separate boxplots for each chromosome. 

Add some horizontal lines showing the limits of the IQR for all genes together.

```{r}
ggplot( mgenes, aes(x= Chr , y= Gene_Length_kb, color=Chr ) ) +  # data and aes
   geom_boxplot() +  # add geom for boxplot
  
  # upper and lower quartiles of the whole dataset
  geom_hline(data = mgenes, 
             aes(yintercept = summary(mgenes$Gene_Length_kb)["3rd Qu."]),
             color="black", linetype="dashed") +
  geom_hline(data = mgenes, 
             aes(yintercept = summary(mgenes$Gene_Length_kb)[2]),
             color="black", linetype="dashed")
```


## 10. Bar charts

# a. Total number of genes per chromosome.

Make a bar chart of the total number of genes on each chromosome.

```{r}
# dplyr
genes_per_chr = mgenes %>%
  group_by(Chr) %>% 
  summarise(ngenes = nrow(End)) %>%
  arrange(nrow(End))
head(genes_per_chr)

# make the plot using `geom_col()`
ggplot( data = genes_per_chr, aes(x = Chr , y = ngenes, fill = Chr  )) +
  geom_col() +
  labs(x="Chromosome", y="Chromosome size (est., in kb)")

```

# b. Size of each chromosome.

Make a bar chart of the approximate size of each chromosome, using the largest gene `End` coordinate for that chromosome as a proxy for chromosome length.

```{r}
# dplyr
chr_size = mgenes %>%
  group_by(Chr) %>% 
  summarise(max_end = max(End)) %>%
  arrange(max_end)
head(chr_size)

# make the plot using `geom_col()`
ggplot( data = chr_size, aes(x = Chr , y = max_end/1000, fill = Chr  )) +
  geom_col() +
  labs(x="Chromosome", y="Chromosome size (est., in kb)")

```

## 11. Make a scatter plot of the total number of genes on each chromosome. Sort them from low to high, and add some nicer labels 

```{r}
# by chr length
chr_length = mgenes %>% group_by(Chr) %>% 
  summarise(chr_size = max(End)) %>% 
  mutate(name=factor(Chr, levels=Chr))

# make the plot
ggplot( data=chr_length, aes(x=reorder(Chr,chr_size),
        y=chr_size, color=Chr)) + 
  geom_point()

```


## 12. SD and CV

Calculate the standard deviation for gene lengths and gene GC-content. 

Then calculate the coefficient of variation by hand, and compare the variation in gene lengths with the variation in gene GC-content. Which one is more variable?

```{r}
# sd


# cv - what is this?


```

