---
title: 'Take-home final: single-cell analysis'
author: "Joseph Berkson [Luok Wen Yong]"
date: "`r format(Sys.time(), '%d %B, %Y')`"
subtitle: XDASI Fall 2021
---

```{r setup, include=FALSE}
# make sure you have these packages installed
library(umap)
library(cluster)
library(dplyr)
library(pheatmap)
library(tidyverse)
```

# Looking under the hood

Single-cell techniques have experienced a real boom in the last several years, and single-cell RNA-seq is progressively becoming as commonplace as "classical" bulk RNA-seq. The computational community has created some tremendous tools that streamline the analyses and make them as user-friendly as possible (think of `Seurat` developed in Satija lab at NYU/NYGC). Here, however, we will do what the majority of non-computational folks do *not* do. We will perform single-cell analysis "by hand", although some steps will be simplified. Why is this useful? First, we will demystify some widely used terms. For example, we will see that a "highly variable" gene and a "marker" gene are very straightforward concepts. Even more importantly, understanding the statistical techniques behind a basic single-cell RNA-seq pipeline will shed light on the limitations of what such analyses can tell us about biology.

Note: if you have previously performed single-cell analyses in Seurat or another package, please do **not** use any of such packages in this exercise!

## 1. Import and clean the data

#### 1.a. Import the counts

Import the count matrix that we provided with this exercise. Determine whether rows or columns contain genes and cells and make sure that you set column and row names correctly.

_Hint: cells are named after the nucleotide sequence of their barcode, and gene names are "MIR1302-10", "FAM138A" etc._

```{r}
singlecell <- read.csv("/Users/lwyong/Desktop/LW Yong NYU/Biostat 2021 Fall/Week 15/TAKE-HOME FINAL attached files Dec 13, 2021 858 PM/XDASI2021-FINAL_data.csv", row.names = "X")

head(singlecell)
```

#### 1.b. Quality control

Remember the homework on Poisson distribution? In droplet-based single-cell sequencing, cell concentration is chosen in such a way that most droplets have no cells (these have already been excluded from our data by read counting software), and those that do mostly contain one cell per droplet. However, there will always be some droplets that contain two or more cells. A simple way to weed out some of them is to discard those cells where the **total** counts are abnormally high. What is abnormal is determined by each researcher individually, and it depends on the biological sample being studied. Our data set includes cells that are similar in size. So, it is safe to assume that they should normally contain similar amount of RNA and cells with unusually high counts are doublets which should be discarded.

+ First, calculate the total number of counts in each cell and plot the distribution.
+ Second, select an arbitrary cutoff, and remove the cells that have more counts than the cutoff from the count matrix. 

_Hint: there should be relatively few outliers in our data. We suggest setting the threshold in a way that less than 1% of cells are discarded._

```{r}
#total no. counts for each column (which is each cell)
singlecell_counts<-colSums(singlecell)
plot(singlecell_counts)

#how many cells in total 
dim(singlecell) #2700

#remove cells with higher counts than 6000
singlecell_filter<-
  singlecell %>% 
  select_if(negate(function(col) sum(col) > 6000))

#double check data
singlecell_filterSUM<-colSums(singlecell_filter)
head(singlecell_filterSUM)
plot(singlecell_filterSUM)


```

## 2. Select highly-variable genes

The most popular methods of single-cell library prep fail to capture all transcript in cells, leading to so-called dropouts, where cells that actually express a gene have 0 read counts for it. Additionally, the sequencing is usually quite shallow (~20 *thousand* reads per cell vs 5-20 *million* reads per sample in bulk RNA-seq). Finally, there is real heterogeneity, even among the cells of the same type and in the same cell state. Together, this makes the count data quite noisy. Therefore, the convention is to only perform most of the downstream analyses with a subset of **highly variable genes**. What does this mean? Generally speaking, variance in gene expression between cells is positively correlated with how highly the gene is expressed: highly expressed genes show higher variance between cells and lowly expressed genes show lower variance between cells. **However, some genes vary between cells more than you would expect just given their expression level.** These genes are called "highly variable" and they have proven to be **the most informative** of transcriptomic differences between cells.

#### 2.a. Compare variance with expression level

+ First, calculate the mean expression level for each gene across cells. 
+ Second, calculate the variance of each gene across cells. 
+ Third, log-transform both measures to equalize the contribution of highly and lowly expressed genes. 

```{r}
#mean across row
singlecell_filter.mean <-rowMeans(singlecell_filter[,1:2674])
str(singlecell_filter.mean)
head(singlecell_filter.mean)
##sanity check for AL627309.1 (if the mean is really 0.0033657442)
sum(singlecell_filter[6,])/2674

#variance across row
library(matrixStats)
singlecell_filter.vars2 <- apply(singlecell_filter[,1:2674], 1, var)
head(singlecell_filter.vars2)
##sanity check for AL627309.1 is var is 0.003355671
var(singlecell_filter[6],)


#log the stuff + add pseudocount
log_sc.filter.mean <-log(singlecell_filter.mean+1)
log_sc.filter.vars <-log(singlecell_filter.vars2+1)

```

_Hint: some genes may not be expressed in any cell, resulting in 0 mean and 0 variance. Log(0) equals + or - infinity, so it is a good idea to perform log(N + 1) transformation instead of simply log(N). This is called "adding a pseudocount"._

+ Finally, plot log-transformed variance on the Y axis and log-transformed mean on the X axis.

```{r}
plot(log_sc.filter.mean,log_sc.filter.vars)
```

#### 2.b. Identify genes with higher variance than predicted by their expression level

Once again, we are interested in finding genes that vary between cells more than we would expect based on their expression level. A simple strategy is to regress expression variance on mean expression and select the data points that deviate from the predicted relationship the most, i.e. the points that have the highest residuals in the regression model. While linear regression is not necessarily the best choice for this, it is good enough for the sake of this exercise.

+ First, regress log-transformed variance on log-transformed mean.
+ Second, extract the residuals from the model.
+ Third, identify the names or indices of the genes with the top 1,000 largest residuals so you can subset them later.

_Hint: there are at least a couple of different ways to do this. You may either `sort()` residuals and extract the names of the largest residuals, which will be corresponding gene names, or `rank()` the residuals and extract the indices of those with the highest ranks._

```{r} 
###USE THIS CHUNK
#linear regression
sc.model2 <- lm(log_sc.filter.vars~log_sc.filter.mean)
summary(sc.model2)
#extract residuals
sc.model.residuals <- sc.model2$residuals
#rank the residuals & extract
top1000 <- sort(sc.model.residuals, decreasing = TRUE)
top1000.index <- top1000[1:1000]
head(top1000.index)
top1000.index2 <-names(top1000.index)[1:1000] 
#subset top1000 data
singlecell_top1000 <-singlecell_filter[top1000.index2,]
```

+ Finally, as a sanity check to make sure your code does what you intended, redraw the plot above but highlight the most highly variable genes by drawing them in a different color.

```{r}
#make df for ggplot
sc.f.mean.var <-cbind(log(singlecell_filter.mean+1),log(singlecell_filter.vars2+1))
sc.f.mean.var.withpseudolog.df<-as.data.frame(sc.f.mean.var)
colnames(sc.f.mean.var.withpseudolog.df)<-c("pseudolog_mean","pseudolog_var")

top1000.pseudolog<-sc.f.mean.var.withpseudolog.df[top1000.index2,]
colnames(sc.f.mean.var.withpseudolog.df)<-c("pseudolog_mean","pseudolog_var")

#the ggplot
ggplot(sc.f.mean.var.withpseudolog.df, 
       mapping = aes(x = pseudolog_mean,
                     y = pseudolog_var))+
  geom_point(col = "black", alpha = 0.3) +
  geom_point(data = top1000.pseudolog,
             aes(x = pseudolog_mean,
                 y = pseudolog_var),
             col = "red", alpha = 0.3)

```

## 3. Pre-process count data

Before performing PCA and other downstream analyses, we need to depth-normalize the counts and transform our data to equalize the contribution of lowly and highly expressed genes. 

+ First, perform CP10K (counts per 10,000 total counts) normalization.

_Hint: you should start by calculating the total number of counts in each cell._

+ Second, log-transform the normalized counts (and remember about what is going to happen to the zeros).
+ Third, subset highly variable genes using the names or indices determined in 2.b. 
+ Fourth, scale and center the log-transformed data such that the mean expression of each gene across cells is ~0 and variance between cells is 1. 

_A very important hint for the last step: you may use `scale()` with default parameters, which will perform centering and scaling in one go. However, be very careful as this function scales columns of the input table and not rows! Ideally, you should build in some sanity checks in your code: after scaling, verify whether the variance of each gene *across cells* really equals 1. If it is not 1, and instead the variance *across genes* equals 1, you need to go back and transpose your data before applying `scale()`._

```{r}
##First, perform CP10K (counts per 10,000 total counts) normalization.
#CP10K = (10^3 * gene reads) / total reads

#total number of count in each cell
sc.normalize1 <- apply(singlecell_filter, 2, function(x) x*10000/sum(x))

#sanity check for normalize B2M (1st cell AAACATACAACCAC.1)
(10000 * singlecell_filter["B2M","AAACATACAACCAC.1"])/colSums(singlecell_filter[1])
sc.normalize1["B2M","AAACATACAACCAC.1"]
#both are equal, good

##Second, log-transform the normalized counts
#log transform
sc.normalize1.log <-log(sc.normalize1+1)

#double check log transformed
log((sc.normalize1["B2M","AAACATACAACCAC.1"])+1)
sc.normalize1.log["B2M","AAACATACAACCAC.1"]

### Third, subset highly variable genes
sc.normalize1.log.top1000<-sc.normalize1.log[top1000.index2,]

###Fourth, scale and center the log-transformed data such that the mean expression of each gene across cells is ~0 and variance between cells is 1. 
#transpose data
sc.normalize1.log.top1000.t<-t(sc.normalize1.log.top1000)
#sclae
sc.normalize1.log.top1000.scaled<-scale(sc.normalize1.log.top1000.t)

#check var ~1 across genes
apply(sc.normalize1.log.top1000.scaled,2,var)

```

## 4. PCA

Since we are dealing with the expression levels of 1,000 genes, dimensionality reduction can help us summarize and visualize the data, as well as determine the major sources of variance. Let us start by performing PCA and examining which genes contribute the most to the first few principal components.

#### 4.a. Perform PCA

+ First, perform PCA on the normalized log-transformed counts of highly variable genes scaled to a mean of 0 and variance of 1 (i.e. on the last thing you created in the previous step). 

_Hint: you may need to transpose your data (yes, again, SORRY!) A useful sanity check is to check the dimensions of the `$x` slot after you have run `prcomp()`: if the number of rows equals the number of cells and the number of columns equals 1,000, all is good. If not, something is wrong and you should try running `prcomp()` on transposed data._

+ Second, plot the proportion of variance explained by the first 20-30 PCs.

_Hint: there are at least two ways of getting the proportion of variance explained by each PC. You may extract the corresponding standard deviation values from `$sdev`, square them and divide each squared value by their sum. Alternatively, you may run `summary()` on your PCA object and extract proportions of variance from the table that is stored in the `$importance` slot of the `summary()` output._

+ Third, plot the data distribution along the first few PCs (decide for yourself which PCs to plot).

```{r}
## First
PCA <- prcomp(sc.normalize1.log.top1000.scaled)
#sanity check
dim(PCA$x)

### Second, plot the proportion of variance explained by the first 20-30 PCs.
variances <- summary(PCA)$importance["Proportion of Variance", 1:20]
plot(variances)

### Third, plot the data distribution along the first few PCs
plot(PCA$x[,1:2])
plot(PCA$x[,2:3])
plot(PCA$x[,3:4])

```

Do you see any structure in the PC plots? How are the data points distributed?

```{r}
#The PC elbow plots plateaus from the 5~6th PCA onwards, meaning that amount of PC should be enough to "explain" the variances in the data. There seems to be about 3 cluster for PC1vsPC2, PC2vsPC3, PC3vsPC4. 
```

#### 4.b. Identify genes with the highest loading on PCs 1 and 2

Find genes that contribute the most to PCs 1 and 2.

_Hint: if you were using the function `princomp()`, variable loadings would be stored in the slot `$loadings` of the output. Somewhat confusingly (for biologists, although not for mathematicians), an analogous slot in the `prcomp()` output is called `$rotation`._

```{r}
PCA_loadings<-as.data.frame(PCA$rotation)

pc1_top10 <- rownames(PCA_loadings[order(abs(PCA_loadings$PC1), decreasing=TRUE),])[1:10]
pc2_top10 <- rownames(PCA_loadings[order(abs(PCA_loadings$PC2), decreasing=TRUE),])[1:10]

pc1_top10
pc2_top10
```

Describe a situation in which you might want to know which genes contribute the most to a given PC axis.

```{r}
# If gene A correlates with PC1, this suggests that gene A (maybe through dose dependant Knockdown) correlates with the variation of the phenotype (you are interested in) of the experiment. 
```

## 5. UMAP and clustering

In many single-cell data sets, there is so much structure in the PC space that it warrants further dimensionality reduction of the PCA output. A popular method to represent multidimensional data in two (or three) dimensions is UMAP, and we can use the function `umap()` from the package `umap` to perform it.

#### 5.a. Perform UMAP

Let us take the main output of PCA (i.e. PC coordinates of all data points) as the input to the `umap()` function. While we can use all 1,000 PCs, it may be more useful to only include those PCs that explain a meaningful amount of the variance.

+ First, look at the plot above which displays the amount of variance explained by each PC and decide how many of them explain more than a "baseline" amount of variance. 
+ Second, use the coordinates of the selected PCs as an input to `umap()` and run the function with default parameters.
+ Third, plot the output.

_Hint: the coordinates of the two UMAP axes are stored in the `$layout` slot of the `umap()` output._

```{r}
#First, look at the plot above which displays the amount of variance explained by each PC
#Second, use the coordinates of the selected PCs as an input to `umap()` and run the function with default parameters.
sc.umap = umap(PCA$x[,c(1:5)],)

#Third, plot the output.
plot(sc.umap$layout)

```

How does the UMAP plot compare to the PCA plots above?

```{r}
#The UMAP plot has 4-ish cluster just like the PCA1.PCA2 clustering above. 
```

#### 5.b. Cluster the UMAP output (`$layout`) using your favorite clustering algorithm

+ First, build diagnostic plots (elbow plot or the plot of average silhouette widths, or both).

_Hint: refer to the last in-class exercise if you need help creating diagnostic plots._

+ Second, decide on the optimum number of clusters and perform clustering with the chosen parameter.

```{r}
###First, build diagnostic plots (elbow plot or the plot of average silhouette widths, or both).

within_vector <- c() # empty vector to store within-cluster variances
between_vector <- c() # empty vector to store between-cluster variances
for (n_centers in 2:20) {
  # perform k-means clustering with k = n_centers
  clustered_km <- kmeans(x = PCA$x[,1:2],
                         centers = n_centers)
  # extract and store within-cluster variance
  within_i <- clustered_km$tot.withinss
  within_vector <- c(within_vector, within_i)
  # extract and store between-cluster variance
  between_i <- clustered_km$betweenss
  between_vector <- c(between_vector, between_i)
}
#The elbow plot
plot(x = 2:20,
     y = within_vector) #5 cluster before level off

### Second, decide on the optimum number of clusters and perform clustering with the chosen parameter.
clustered_km = kmeans(x = PCA$x[,1:2],
                      centers = 5)
plot(PCA$x[,1:2],
     col = clustered_km$cluster)
plot(sc.umap$layout,
     col = clustered_km$cluster)

```

How happy are you with the clustering results? What issues did you encounter?

```{r}
#The UMAP seems to suggests that 3~4 ish cluster is sufficient. This means we can lessen the PC used. With UMAP there seems to be one outlier. 
```

## 6. Marker genes

Cell type markers, or cluster markers, are nothing else than genes that are specifically expressed in a certain cluster or in a subset of clusters. They are determined by performing differential expression testing with some semi-arbitrary filtering applied afterwards. As for differential expression testing, you may be shocked to find out that it is predominantly performed by doing a bunch of Wilcoxon tests (i.e. the simplest non-parametric test and not some fancy DESeq2-style modeling). A Wilcoxon test is done for each gene, where you compare expression values in cells that belong to a given cluster vs. the rest of the cells.

#### 6.a. Identify markers for cluster 1

+ First, for each gene, split the **normalized and log-transformed** expression values into "cells in cluster 1" and "cells NOT in cluster 1" and calculate 
(a) log2 fold change and 
(b) Wilcoxon test p-value between the two groups. 
+ Second, perform FDR correction of the p-values. 
+ Third, filter the genes by significant adjusted p-values and sort them by log2 fold change.

_Hint: you may also want to remove genes with + or - infinity log2 fold changes._

```{r}
#First, for each gene, split the **normalized and log-transformed** expression values into "cells in cluster 1" and "cells NOT in cluster 1" and calculate (a) log2 fold change and (b)

cluster1 <- clustered_km$cluster[clustered_km$cluster == 1]
head(cluster1)
cluster1_NOT <- clustered_km$cluster[clustered_km$cluster != 1]
head(cluster1_NOT)

gene.cluster1<-sc.normalize1.log[,names(cluster1)]
gene.NOTcluster1<-sc.normalize1.log[,names(cluster1_NOT)]
dim(gene.cluster1)
dim(gene.NOTcluster1)

#(a) log2 fold change
#fold change
gene.cluster1.mean <- rowMeans(gene.cluster1) 
gene.NOTcluster1.mean <- rowMeans(gene.NOTcluster1)
gene.cluster1.NOTcluster1.mean <-cbind(gene.cluster1.mean,gene.NOTcluster1.mean)

#log2
gene.cluster1.NOTcluster1.mean.log2<- transform(gene.cluster1.NOTcluster1.mean, log2foldchange = log2((gene.cluster1.mean/gene.NOTcluster1.mean)))

#(b) wilcoxon test
data_combined <- data.frame("p_values" =  rep(NA, 32738))
for (i in 1:nrow(data_combined)) {
  pvalue = wilcox.test(gene.cluster1[i,],gene.NOTcluster1[i,])$p.value
  data_combined$p_values[i] = pvalue
}
head(data_combined)

#combine means, fold change, log2, and p values from wilcox
gene.cluster1.NOTcluster1.mean.log2$p.value.wilcox <- data_combined$p_values
#remove NA & infinity
final.combine <- gene.cluster1.NOTcluster1.mean.log2 %>% mutate_all(~ifelse(is.nan(.), NA, .))
final.combine <- final.combine %>% mutate_all(~ifelse(is.infinite(.), NA, .))
final.combine <- final.combine %>% drop_na() 

### Second, perform FDR correction of the p-values. 
final.combine$pvalues_adj_r <- 
  p.adjust(final.combine$p.value.wilcox,method = "fdr")

### filter
sig_p.values.adjusted <- final.combine %>% 
  filter(pvalues_adj_r < 0.05) %>%  
  arrange(desc(log2foldchange))

```

#### 6.b. Plot expression of the top markers across clusters

Let us create a heatmap with the average expression of selected markers in each cluster. 

+ First, select several (e.g. 10) top genes from the filtered list of cluster 1 markers above. 
+ Second, for each of these genes, average the **normalized and log-transformed** counts between all cells of each cluster. 

_Hint: you may find it useful to arrange the data as a matrix or data.frame with clusters as rows and genes as columns. Each cell should then contain an average expression value for a given gene in the given cluster._ 

+ Third, plot the data on a heatmap.

_Hint: base R has the function `heatmap()`, but `pheatmap()` from the package `pheatmap` is much nicer._

```{r}
###First, select several (e.g. 10) top genes from the filtered list of cluster 1 markers above.

#transpose data - cell as row and gene as column now
sc.normalize1.log_t<-as.data.frame(t(sc.normalize1.log))
# add the cluster number for each cell (clustered_km has this info of which cell is in which cluster)
sc.normalize1.log_t$cluster = clustered_km$cluster
# transpose back
sc.normalize1.log_t<-as.data.frame(t(sc.normalize1.log_t))

#find the top10 for cluster 1
top10.clus1.index <- rownames(sig_p.values.adjusted[c(1:10),])
print(top10.clus1.index)

###Second, for each of these genes, average the **normalized and log-transformed** counts between all cells of each cluster.

#prep data
sc.normalize1.log.top10 <- sc.normalize1.log_t[top10.clus1.index,]
top10gene_diffcluster = rbind(sc.normalize1.log.top10 , sc.normalize1.log_t[32739, ])
head(top10gene_diffcluster)

#prep data for heatmap
library(reshape2)
top10gene_diffcluster_t = as.data.frame(t(top10gene_diffcluster))
top10gene_long = melt(top10gene_diffcluster_t, id.vars =c("cluster"), value.name = "count", variable.name = "gene") 
head(top10gene_long)
dim(top10gene_long)

top10gene_long_summary= top10gene_long %>% group_by(gene,cluster) %>% summarize(average = mean(count))

#converted to wide for heatmap
top10gene_wide_summary = spread(top10gene_long_summary, gene, average)
top10gene_wide_summary = as.data.frame(top10gene_wide_summary)
rownames(top10gene_wide_summary) = top10gene_wide_summary$cluster
top10gene_wide_summary= select(top10gene_wide_summary, -1)

head(top10gene_wide_summary)

#without scale
pheatmap(top10gene_wide_summary, main = "no scale")
#with scale 
pheatmap(scale(top10gene_wide_summary), main = "scaled")

```

Does the heatmap pattern correspond to your expectations? Try scaling the data (hint: use parameter `scale` inside `pheatmap()`). Was it useful? How did it affect your interpretation of the results?

```
Yes, we expect that all these genes should be highly represented in cluster 1, which appears as red in the heatmap. Rescaling was helpful as it really highlights the presence of those 10 genes in cluster 1. 
```

# THE END! ENJOY YOUR WINTER BREAK!
