---
title: "XDAS 2020 HW6"
author: "Venkatraman Ramakrishnan"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE)
```

## Review

+ Class notes:
  + [Chi-Square](https://xdas.bio.nyu.edu/topics/week_06/1_Chi-square/Chisq.html)
  + [Contingency Analysis](https://xdas.bio.nyu.edu/topics/week_06/2.1_Association_RR_OR_Fishers/ContingencyAnalysis_2020.html)
+ W&S, Chapters 8 and 9

Optional:

+ [Aho, Chapter 11](https://drive.google.com/open?id=1FIldd7NHczH-6HMAdIkU45aIuYh00uhA)
+ [Dalgaard, Chapter 7](https://drive.google.com/open?id=1Tmi38WptGxu0xvsWanQ3n_q2fa_2ui9i)

## Q0: Load BMI Data

This is a published data set which has measurements of Body Mass Index (BMI) and the presence or absence of several single-nucleotide polymorphisms (SNPs), as detected by genotyping on the same individuals. 
There is data from 79 people. Each person has a recorded BMI (numeric) and the presence or absence of 7 SNPs (encoded as a factor with Y and N)

Our goal in this study is to determine whether any of the SNPs are associated with Obesity.

#### Load the `HW6_SNP_Data.tsv` file using `read.delim()` function and call the variable `hw6.data`.

```{r}
hw6.data <- read.delim("HW6_SNP_Data.tsv")
str(hw6.data)
```

#### Briefly describe the structure of this observational data set. What measurements have been taken, and what is the data type of those measurements?

```
<!-- INSERT ANSWER HERE -->
```


## Q1: $\chi^2$ Test

#### a) Add a column `Overweight` to the `hw6.data` and set the value to `TRUE` if the BMI is equal to greather than 25 and `False` if not. Individuals will be considered `Overweight` if they have a BMI of 25 or more.

```{r}
hw6.data$Overweight <- hw6.data$BMI >= 25
str(hw6.data)
```

#### How many individuals are overweight by this definition? How many are not overweight?

```{r}
table(hw6.data$Overweight)
```

#### b) Generate a contingency table showing Overweight and the presence of SNP1.
**Hint - you may choose to use the table() function**
```{r}
snp1.contingency.table <- table(hw6.data[, c("Overweight", "SNP1")])
snp1.contingency.table
```

#### c) Generate an expectation table showing the EXPECTED number of individuals when comparing Overweight to SNP1. Calculate this table using the formula discussed in class.
$$E_{A{\cap}B} = Pr_A * Pr_B * Total$$

```{r}

# The totals
obesity = table(hw6.data$Overweight)
snp1present=table(hw6.data$SNP1)

expected_snp1 = matrix(NA, nrow=2,ncol=2)
expected_snp1[1,1]=obesity[1]*snp1present[1]/sum(obesity)
expected_snp1[2,1]=obesity[2]*snp1present[1]/sum(obesity)
expected_snp1[1,2]=obesity[1]*snp1present[2]/sum(obesity)
expected_snp1[2,2]=obesity[2]*snp1present[2]/sum(obesity)
rownames(expected_snp1)=c("False","True")
colnames(expected_snp1)=c("N","Y")
```

```{r}
print(expected_snp1)
```
#### d) Calculate the $\chi^2$ statistic

```{r}
snp1_chisq_statistic = sum(((snp1.contingency.table - expected_snp1)**2)/expected_snp1)
snp1_chisq_statistic
```


#### e) Test your observed contingency table for Overweight vs SNP1 with the Chi-Squared test. Use the `pchisq()` function with the $\chi^2$ test calculated in the previous step.

```{r}
pchisq(snp1_chisq_statistic, 1, lower.tail = F)

```
#### f) Confirm your results using the `chisq.test()`
Be sure to turn the correction off since we did not perform the correction in the calculation above.

```{r}

chisq.test(snp1.contingency.table, correct=F)

```

#### g) What is the null hypothesis for this chi-squared test?

```
<!-- INSERT ANSWER HERE -->

The two variables (Overweight & SNP1) are independent
```

#### h) Interpret the result of your chi-square test between the Overweight trait and SNP1.

```
<!-- INSERT ANSWER HERE -->

We fail to reject the null hypothesis, and it appears that SNP1 is unlikely to have an effect on obesity 
```
## Q2: Relative Risk vs Odds Ratio

#### a) Calculate the **relative risk** for SNP1 and Obesity.

```{r}
#prob of having SNP1 in obese / prob of having snp1 in not obese

risk1=snp1.contingency.table[1,2]/sum(snp1.contingency.table[1,])
risk2=snp1.contingency.table[2,2]/sum(snp1.contingency.table[2,])

relrisk = risk2/risk1
relrisk
```

#### b) Confirm your results using the `riskratio` function in the `epitools` package.


```{r}

library("epitools")
riskratio(snp1.contingency.table,correction = F)
```

#### c) Calculate the **odds ratio** ratio for SNP1 and Obesity.

```{r}
# odds ratio is the ratio of the odds
# odds that a obese person has snp1 is
odds1= snp1.contingency.table[1,2] / snp1.contingency.table[1,1] 

odds2 =snp1.contingency.table[2,2]/ snp1.contingency.table[2,1]

odd_ratio = odds2/odds1

odd_ratio
```

#### d) Confirm your results using the `oddsratio` function in the `epitools` package.

Make sure you use the `wald` method in the `oddsratio()` function.

```{r}

library("epitools")
oddsratio(snp1.contingency.table, method="wald")
```

## Q3: Fisher's exact test

#### a) Test the relationship between the Overweight trait and SNP1 with a Fisher's Exact Test

```{r}
snp1.fisher <- fisher.test(snp1.contingency.table)
print(snp1.fisher)

```

#### b) Describe the difference between the Fisher Exact Test and the Chi-Squared test

<!-- INSERT ANSWER HERE -->

The fisher exact test is nonparametric and calculates p-values based on the hypergeometric distribution. The chi-square test is parametric; chi is calculated as (observed - expected) ** / expected. P-values are calculated based on the chi distribution.

#### c)What is the null hypothesis for this Fisher's Exact Test?

<!-- INSERT ANSWER HERE -->

The two variables are independent

#### d) Interpret the result of your Fisher's Exact test. Compare the p-values to the chi-squared test.

<!-- INSERT ANSWER HERE -->


#### e) Check the other SNPs to see which SNP is most associated with obesity.

You may use the Fisher's eExact tTest to check for this.


```{r}
for (i in 2:ncol(hw6.data)) {
  print(colnames(hw6.data)[i])
  snp.contingency.table <- table(hw6.data[, c("Overweight", colnames(hw6.data)[i])])
  snp.fisher <- fisher.test(snp.contingency.table)
  print(snp.fisher$p.value)
}


```

Are any of the SNPs significantly asociated with BMI?

```
<!-- INSERT YOUR ANSWER -->
```