---
title: 'Homework 4: Probability'
subtitle: "XDASI Fall 2021"
author: "Sir Ronald Fisher [YOUR NAME HERE]"
date: "Due: 9/30/2021"
output:
  html_document:
    toc: 3
    toc_float: TRUE
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE)
library(ggplot2)
library(gmodels)
# remove "error = TRUE" to make knitr halt on errors
# (do this when you are ready to check if your code will knit without errors)
```

# 1. Probability density and cumulative probability

You are screening for genetic interactions between genes in yeast that cause deleterious fitness effects. To establish a baseline for your experiments, you decide to first estimate the probability density of colony diameters (in mm) for a population of WT yeast grown in controlled conditions (48 hours at 37°C). Fortunately, you have an automated image recognition system that allows you to do this for a large number of individual colonies!

#### a. Simulate a sample of 1000 measured colonies with an average diameter of 1mm and a standard deviation of 0.1mm and store it in a variable. Set a seed to make your work reproducible. 

What are the actual mean and SD? What about the IQR? Draw a boxplot to get a feel for the sample distribution.

```{r}
# your answer here
seed = 123456789
set.seed(seed)
random_yeast_sample = rnorm(1000, 1, 0.1)
mean(random_yeast_sample)
IQR(random_yeast_sample)
boxplot(random_yeast_sample)

```

#### b. What are the SEM and 95%CI for this sample? How are these summary statistics computed, and what do they mean?

```{r}
# Option 1: SEM and CI by hand

# Option 2: using gmodels package
total_ci = ci(random_yeast_sample)
sem = total_ci['Std. Error']
sem
yeast_ci = c(total_ci['CI lower'], total_ci['CI upper'])
yeast_ci

# The standard error of the mean is calculated as the standard deviation of the sample divided by the square root of the sample size. The 95% CI is calculated as approximately the mean of the sample +/- two times the standard error of the mean. The 95% CI represents an interval in whcih we are 95% certain that the true population mean falls. The standard error of the mean is the standard deviation of the distribution of the sample mean
```


#### c. Draw a histogram of the sampled data, and make a qqplot to confirm that the sample really comes from a normal distribution.

***Extra credit:*** *See if you can overlay some additional information on the histogram, such as vertical lines showing some statistics, a smoothened density line for your sample distribution, and/or an "ideal" normal curve with the same parameters.*

```{r}
# Histograms
yeast_df = data.frame(random_yeast_sample)
colnames(yeast_df) = c('Size')
yeast_df

# Option 1: base R

# Option 2: ggplot

bw = 0.01

ggplot(data = yeast_df,
       mapping = aes(x = Size)) + 
  
  geom_histogram(binwidth=bw, color='black', fill='aquamarine')



# qqplot
qqnorm(yeast_df$Size)
qqline(yeast_df$Size)

```


Use the R functions for the CDF (`pnorm()`) and inverse CDF (`qnorm()`) to answer questions d-f below.

#### d. What is the total probability that a randomly sampled colony from the parent population measures between 0.93 and 1.07 mm?

```{r}
# your answer here
yeast_sd = sd(yeast_df$Size)
yeast_mean = mean(yeast_df$Size)
pnorm(1.07, mean=yeast_mean, sd=yeast_sd) - pnorm(0.93, mean=yeast_mean, sd=yeast_sd) 
```


#### e. What is the proportion of data that are <1 or >1 standard deviations away from the mean?

```{r}
# your answer here
pnorm(yeast_mean - yeast_sd, mean=yeast_mean, sd=yeast_sd) + (1 - pnorm(yeast_mean + yeast_sd, mean=yeast_mean, sd=yeast_sd))
```


#### f. What diameters correspond to the 10% and 90% quantiles?

```{r}
# your answer here
qnorm(0.1, mean=yeast_mean, sd=yeast_sd)
qnorm(0.9, mean=yeast_mean, sd=yeast_sd)
```


# 2. Co-occurrence of infection

Among women voluntarily tested for sexually transmitted diseases in one university (modified from Tábora et al. 2005):

+ 18% tested positive for human papilloma virus (HPV) only,
+ 2% tested positive for Chlamydia only, and
+ 3% tested positive for both HPV and Chlamydia. 

#### a. What is the total frequency of each STD in the population?

```{r}
# given probabilities
p.hpv = 0.18
p.chl = 0.02
p.hpv.chl = 0.03

p.hpv.tot = 0.21
p.chl.tot = 0.05

# The total frequency of hpv is 21% and the frequency of chlamydia in the population is 5%
```


For the next 3 questions, use set notation to write out the question in terms of set operations (as a comment) and then compute the answer. You may write probabilities using the following syntax, e.g.:

+ P(A) * P(B)
+ P(A) + P(B)
+ P(A AND B)
+ P(A OR B)
+ P(A | B)
+ etc.

#### b. What is the probability that a woman will test positive for Chlamydia, given that she is positive for HPV?

```{r}
# notation
#a is testing positive for chlamydia b is testing positive for hpv

#P(A | B) * P(B) = P(A AND B) thus P(A | B) = P(A AND B) / P(B)

# computation

p.chl.given.hpv = p.hpv.chl / p.hpv.tot
p.chl.given.hpv
```


#### c. What is the probability that a woman who tests positive for Chlamydia will also test positive HPV?

```{r}
# notation

#P(B | A) * P(A) = P(A AND B) thus P(B | A) = P(A AND B) / P(A)

# computation
p.hpv.given.chl = p.hpv.chl / p.chl.tot
p.hpv.given.chl
```


#### d. What is the total probability that a woman from this population who gets tested would test positive for either HPV and/or Chlamydia?

```{r}
# notation
#This is asking for P(A) + P(B) - P(A and B)
p.hpv.tot + p.chl.tot - p.hpv.chl
# computation
```


#### e. Is the occurrence of these infections independent? Explain.

```{r}
# No as the probability of both occurring is not the same as P(A) * P(B)
```


# 3. Allele frequencies

Many gene loci have a major allele and a number of minor alleles in the population. Let's say there are 5 alleles for a particular locus, that individuals mate randomly with respect to the genotype at this locus, and that allele frequencies for A1-A5 are 0.82, 0.06, 0.05, 0.04, and 0.03 respectively.

For each of the questions below, write out the probability equations (e.g. P(Ai) for each term) and then compute the answers numerically.

#### a. What is the probability that any single allele chosen at random from the population is either A2 or A5? Write out the probability statement as a comment below and then computer the answer.

```{r}
# Allele frequencies
A1 = 0.82
A2 = 0.06
A3 = 0.05
A4 = 0.04
A5 = 0.03

# P(A2) + P(A5)

# computation
A2 + A5

```


#### b. What is the probability that an individual carries two A3 alleles?

```{r}
# P(A3) * P(A3)

# computation

A3^2
```


#### c. What is the probability that someone does NOT carry two A3 alleles?

```{r}
# 1 - P(A3 and A3)

# computation

1 - A3^2
```


#### d. What is the probability that someone is heterozygous for A1 and A4?

```{r}
# Assuming this means someone being A1/A4

#P(A1) * P(A4)

A1 * A4
```


#### e. What is the probability that *neither* of two random individuals in the population would carry two A1 alleles?

*Hint: first consider (i) the probability of carrying two A1 alleles, then (ii) the probability of NOT having two A1 alleles, and finally (iii) the probability that 2 individuals are not carrying two A1 alleles.*

```{r}
# prob of carrying two A1 alleles

A12 = A1 * A1

# prob of NOT having two A1 alleles

notA1 = 1 - A12

# P(2 individuals are NOT A1,A1)

1 - (notA1) ^ 2

```


#### f. *Extra credit:* What is the probability that 3 random individuals carry no A3 or A4 alleles at all? (Remember that each individual carries two alleles.) 

```{r}
# prob of an allele being A3 or A4

A3or4 = A3 + A4

# prob of an allele not being A3 or A4

notA3or4 = 1 - A3or4

# P(neither of 2 alleles is A3 or A4)

both_not_A3or4 = norA3or4 ^ 2

# P(3 people have no A3 or A4 alleles)

both_not_A3or4 ^ 3

```


# 4. Trait inheritance

Some people are hypersensitive to the smell of asparagus, and can even detect a strong odor in the urine of a person who has recently eaten asparagus. This trait turns out to have a simple genetic basis (Online Mendelian Inheritance in Man, 2019):

+ An individual with one or two copies of the $A$ allele at the gene ($AA$ or $Aa$ genotypes) can smell asparagus in urine, whereas 
+ a person with two copies of the alternative $a$ allele ($aa$ genotypes) cannot. 

**In the human population, 5% of alleles are $A$ and 95% are $a$.**

Assume that men and women in the population have the same allele frequencies at the asparagus-smelling gene and that marriage and child production are independent of the genotype at the gene.


#### a. What is the probability that a randomly sampled individual from the population has two copies of the $a$ allele (that is, that it has an $aa$ genotype)? 

*Note: below we suggest to use a list to hold all of the different variables you will need to calculate, to help you keep track of the combinations. A list uses similar syntax to a data frame, except it can hold any kind of data of any size.*

```{r}
Pr=list()
Pr$a=0.95
Pr$A=0.05

Pr$aa = 0.95 * 0.95
Pr$aa
```


#### b. What is the probability that both members of a randomly sampled married couple (man and woman) are $aa$ at the asparagus-smelling gene? 

```{r}
Pr$aa_aa = Pr$aa ^ 2
Pr$aa_aa
```


#### c. What is the probability that both members of a randomly sampled married couple (man and woman) are heterozygotes at this locus (meaning that each person has one allele $A$ and one allele $a$)? Remember that each allele can be inherited from either parent.

```{r}
# first calculate probability that one individual is heterozygous
Pr$Aa = Pr$A * Pr$a
Pr$Aa

# probability that both parents are heterozygous
Pr$Aa_Aa = Pr$Aa ^ 2
Pr$Aa_Aa        
```


#### d. What is the conditional probability that the first child of the couple in (c) is an $aa$ homozygote? Remember that the child must receive exactly one allele from each parent.

***Beware: here you need to consider transmission probabilities, not the allele frequencies in the population! Think "Punnett square".***

```{r collapse=TRUE}
# conditional probability that Aa x Aa -> aa (I think it should be aa not Aa is it was originally)
Pr$Caa.FAa_MAa = 0.5 * 0.5
Pr$Caa.FAa_MAa
```


#### e. Above we figured out the probability that a child of known parentage inherits a particular allele combination. What if we want to turn this around and ask, ***If a randomly sampled child has the $aa$ genotype, what is the probability that both its parents were both $Aa$ heterozygotes?***

To answer this we need to use **Bayes' Theorem**. Recall the definition of conditional probability:

$$ P(F_{Aa} M_{Aa} | C_{aa}):= \frac{P(F_{Aa} M_{Aa} \cap C_{aa}) }{ P(C_{aa}) }$$
and

$$ P(C_{aa} | F_{Aa} M_{Aa}):= \frac{P(F_{Aa} M_{Aa} \cap C_{aa}) }{ P(F_{Aa} M_{Aa}) }$$


Therefore, the intersection (co-occurrence) of the parental and child alleles can be written equivalently as:

$$P(F_{Aa} M_{Aa} \cap C_{aa}) = P(F_{Aa} M_{Aa} | C_{aa}) * P(C_{aa}) = P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})$$

Fantastic! Now we have what we need to find the desired ***posterior probability*** $P(F_{Aa} M_{Aa} | C_{aa})$, given the known

+ ***priors*** $P(F_{alleles}M_{alleles})$ and
+ ***likelihoods*** $P(C_{aa}|F_{alleles}M_{alleles})$.

All we have to do is rearrange this equation and use Bayes Theorem:

$$P(F_{Aa} M_{Aa} | C_{aa}) = \frac{P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})}{P(C_{aa})} = \frac{P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})}{\sum_{ij} P(C_{aa}|F_iM_j) * P(F_iM_j)}$$

where $i,j \in \{AA, Aa, aA, aa\}$. 

This equation represents ***the fraction of the time that we expect to see one particular combination of parental alleles given an $aa$ child (i.e. both parents are $Aa$ heterozygotes), relative to all the possible ways a child could inherit two $a$ alleles.***

To compute the ***posterior probability***, we need to first quantify a bunch of individual probabilities, so let's break the problem into pieces and compute the required components in stages:

- **i. Probabilities of each allele combination in any individual in the population (parental frequencies):**

```{r collapse = TRUE}
# probabilities of each allele combination in the parents
Pr$AA =  Pr$A ^ 2
Pr$Aa =  Pr$A * Pr$a
Pr$aA =  Pr$A * Pr$a
Pr$aa =  Pr$a ^ 2

Pr$AA
Pr$Aa
Pr$aA
Pr$aa
```

- **ii. *Priors* for each combination of parental alleles (states of nature):**

```{r collapse=TRUE}
# priors for each combination of parental alleles
Pr$Faa_Maa = Pr$aa ^ 2
Pr$FAa_Maa = Pr$Aa * Pr$aa
Pr$FAA_Maa = Pr$AA * Pr$aa
Pr$Faa_MAa = Pr$aa * Pr$Aa
Pr$FAa_MAa = Pr$Aa ^ 2
Pr$FAA_MAa = Pr$AA * Pr$Aa
Pr$Faa_MAA = Pr$aa * Pr$AA
Pr$FAa_MAA = Pr$Aa * Pr$AA
Pr$FAA_MAA = Pr$AA ^ 2

Pr$Faa_Maa
Pr$FAa_Maa
Pr$FAA_Maa
Pr$Faa_MAa
Pr$FAa_MAa
Pr$FAA_MAa
Pr$Faa_MAA
Pr$FAa_MAA
Pr$FAA_MAA
```

- **iii. *Likelihoods* (conditional probabilities) for $C_{aa}$ given each combination of parental alleles:**

```{r}
# individual likelihoods (think Punnett square)
Pr$Caa.Faa_Maa = 1
Pr$Caa.FAa_Maa = 0.5
Pr$Caa.FAA_Maa = 0
Pr$Caa.Faa_MAa = 0.5
Pr$Caa.FAa_MAa = 0.25
Pr$Caa.FAA_MAa = 0
Pr$Caa.Faa_MAA = 0
Pr$Caa.FAa_MAA = 0
Pr$Caa.FAA_MAA = 0
```


- **iv. *Total probability* $Pr(C_{aa})$ that the child is $aa$ (the denominator):**

To find this, unfortunately need to consider the probabilities of all possible combinations of parental allele contributions, or possible ***states of nature***:

$$\begin{aligned}P(C_{aa}) =& P(C_{aa}|F_{aa}M_{aa}) * P(F_{aa}M_{aa})  \ + \\
                            & P(C_{aa}|F_{Aa}M_{aa}) * P(F_{Aa}M_{aa})  \ + \\
                            & P(C_{aa}|F_{AA}M_{aa}) * P(F_{AA}M_{aa})  \ + \\
                            & P(C_{aa}|F_{aa}M_{Aa}) * P(F_{aa}M_{Aa})  \ + \\
                            & P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})  \ + \\
                            & P(C_{aa}|F_{AA}M_{Aa}) * P(F_{AA}M_{Aa})  \ + \\
                            & P(C_{aa}|F_{aa}M_{AA}) * P(F_{aa}M_{AA})  \ + \\
                            & P(C_{aa}|F_{Aa}M_{AA}) * P(F_{Aa}M_{AA})  \ + \\
                            & P(C_{aa}|F_{AA}M_{AA}) * P(F_{AA}M_{AA}) 
\end{aligned}$$


```{r}
# total probability P(C_Aa)
Pr$Caa = Pr$Caa.Faa_Maa * Pr$Faa_Maa +
         Pr$Caa.FAa_MAa * Pr$FAa_MAa +
         Pr$Caa.FAa_Maa * Pr$FAa_Maa +
         Pr$Caa.FAA_Maa * Pr$FAA_Maa +
  Pr$Caa.Faa_MAA * Pr$Faa_MAA +
  Pr$Caa.FAa_MAA * Pr$FAa_MAA +
  Pr$Caa.FAA_MAA * Pr$FAA_MAA# first line filled in for you ... add the rest below
         

Pr$Caa
```


- **v. Finally, we can plug the desired variables from above into the original Bayes equation to find the *posterior probability:* **

$$P(F_{Aa}M_{Aa} | C_{aa}) = \frac{P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})}{P(C_{aa})}$$

```{r}
# Posterior probability of P(F_Aa, M_Aa | C_aa):
Pr$FAa_MAa.Caa = (Pr$Caa.FAa_MAa * Pr$FAa_MAa) / Pr$Caa
Pr$FAa_MAa.Caa
```


#### f. If a randomly sampled child has the $aa$ genotype, what is the probability that both its parents were also $aa$?

The ***posterior probability*** $P(F_{aa} M_{aa} | C_{aa})$ is:

$$P(F_{aa} M_{aa} | C_{aa}) = \frac{P(C_{aa}|F_{aa}M_{aa}) * P(F_{aa}M_{aa})}{P(C_{aa})} = \frac{P(C_{aa}|F_{aa}M_{aa}) * P(F_{aa}M_{aa})}{\sum_{ij} P(C_{aa}|F_iM_j) * P(F_iM_j)}$$

We computed all the relevant quantities above, so now this should be easy!

```{r}
Pr$Faa_Maa.Caa = (Pr$Caa.Faa_Maa * Pr$Faa_Maa) / Pr$Caa
Pr$Faa_Maa.Caa
```

---