---
title: 'Homework 4: Probability'
subtitle: "XDASI Fall 2021"
author: "Mari Shiratori"
date: "Due: 9/30/2021"
output:
  pdf_document: default
  html_document:
    toc: 3
    toc_float: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = FALSE)
library(ggplot2)
library(gmodels)
# remove "error = TRUE" to make knitr halt on errors
# (do this when you are ready to check if your code will knit without errors)
```

# 1. Probability density and cumulative probability

You are screening for genetic interactions between genes in yeast that cause deleterious fitness effects. To establish a baseline for your experiments, you decide to first estimate the probability density of colony diameters (in mm) for a population of WT yeast grown in controlled conditions (48 hours at 37°C). Fortunately, you have an automated image recognition system that allows you to do this for a large number of individual colonies!

#### a. Simulate a sample of 1000 measured colonies with an average diameter of 1mm and a standard deviation of 0.1mm and store it in a variable. Set a seed to make your work reproducible. 

What are the actual mean and SD? What about the IQR? Draw a boxplot to get a feel for the sample distribution.

```{r}
# set a seed
set.seed(0)

# sample
sample = 1000
distribution = rnorm(sample, mean = 1, sd = 0.1)
mean(distribution)

# mean and sd
# The actual mean is 0.998417
sd(distribution)
# The actual sd is 0.09980048

# IQR and boxplot
summary(distribution)
# The 1st quartile is .9292 and the 3rd quartile is 1.0688

boxplot(distribution)

```

#### b. What are the SEM and 95%CI for this sample? How are these summary statistics computed, and what do they mean?

```{r}
# Option 1: compute by hand

#Calculating standard error
sem_distribution <- sd(distribution)/sqrt(sample)
sem_distribution

error <- qnorm(0.975) * sem_distribution
error

#Calculating lower interval
lower_bound <- mean(distribution) - error
lower_bound

#Calculating upper interval
upper_bound <- mean(distribution) + error
upper_bound

# Option 2: using gmodels package
library(gmodels)
ci(distribution)

# Explanation
# The SEM is computed by taking the standard deviation of the entire distribution which in this case is the 1000 values that were simulated using rnorm in the previous code block and dividing by the square root of the sample number, which in this case is 1000.
# To calculate the confidence intervals, the error for a 95% interval was calculated by taking the Qnorm of 0.975 and multiplying by the SEM calculated before and then adding or subtracting that value from the mean of the distribution (in this case around 0.9984 as we saw from the code block above) to get the upper and lower bound of the 95% confidence interval respectively
# The gmodels results and the results I got more manually are very similar.
```


#### c. Draw a histogram of the sampled data, and make a qqplot to confirm that the sample really comes from a normal distribution.

***Extra credit:*** *See if you can overlay some additional information on the histogram, such as vertical lines showing some statistics, a smoothened density line for your sample distribution, and/or an "ideal" normal curve with the same parameters.*

```{r}
# Histograms

# Option 1: base R
hist(distribution)

# Option 2: ggplot
#turn distribution into a data frame so ggplot can read it
distribution_df <- as.data.frame(distribution)

ggplot(data = distribution_df,
       aes(x = distribution)) +
  geom_histogram(binwidth = 0.05, fill = "mistyrose2", color = "grey25") +
  geom_vline(aes(xintercept = mean(distribution)),
             color = "deeppink4",
             size = 1,
             linetype = "dashed") +
  geom_vline(aes(xintercept = summary(distribution)[2]),
             color = "slateblue4",
             size = 1,
             linetype = "dotted") +
  geom_vline(aes(xintercept = summary(distribution)[5]),
             color = "mediumseagreen",
             size = 1,
             linetype = "dotted") +
  geom_text(aes(x = 1.03, 
                y = 210, 
                label = "Mean"), 
            color = "deeppink4") +
  geom_text(aes(x = 0.85,
                y = 210,
                label = "1st Quartile"),
            color = "slateblue4") +
  geom_text(aes(x = 1.15,
                y = 210,
                label = "3rd Quartile"),
            color = "mediumseagreen")


#Unsure how to plot an ideal normal curve onto this histogram.

# qqplot
qqnorm(distribution_df$distribution)
qqline(distribution_df$distribution)
```


Use the R functions for the CDF (`pnorm()`) and inverse CDF (`qnorm()`) to answer questions d-f below.

#### d. What is the total probability that a randomly sampled colony from the parent population measures between 0.93 and 1.07 mm?

```{r}
pnorm(1.07, mean = 1, sd = 0.1, lower.tail = TRUE) - pnorm(0.93, mean = 1, sd = 0.1, lower.tail = TRUE)
# Your probability of this is about 52%
```


#### e. What is the proportion of data that are <1 or >1 standard deviations away from the mean?

```{r}
pnorm(1.1, mean = 1, sd = 0.1, lower.tail = TRUE) - pnorm (0.9, mean = 1, sd = 0.1, lower.tail = TRUE)

# around 68% which tracks with what we expect in a normal distribution. 
# I am taking this question to mean what is the proportion of data within one standard deviation of the mean.
```


#### f. What diameters correspond to the 10% and 90% quantiles?

```{r}
qnorm(0.1, mean = 1, sd = 0.1, lower.tail = TRUE)
qnorm(0.9, mean = 1, sd = 0.1, lower.tail = TRUE)

#In the 10% quantile is a diameter of around 0.87mm and in the 90% quantile is a diameter around 1.13 mm.
```


# 2. Co-occurrence of infection

Among women voluntarily tested for sexually transmitted diseases in one university (modified from Tábora et al. 2005):

+ 18% tested positive for human papilloma virus (HPV) only,
+ 2% tested positive for Chlamydia only, and
+ 3% tested positive for both HPV and Chlamydia. 

#### a. What is the total frequency of each STD in the population?

```{r}
# probabilities
p.hpv = 0.18
p.chl = 0.02
p.hpv.chl = 0.03

# total probabilities
p.hpv.tot = p.hpv + p.hpv.chl
p.chl.tot = p.chl + p.hpv.chl

p.hpv.tot
p.chl.tot
```


For the next 3 questions, use set notation to write out the question in terms of set operations (as a comment) and then compute the answer. You may write probabilities using the following syntax, e.g.:

+ P(A) * P(B)
+ P(A) + P(B)
+ P(A AND B)
+ P(A OR B)
+ P(A | B)
+ etc.

#### b. What is the probability that a woman will test positive for Chlamydia, given that she is positive for HPV?

```{r}
# set operations
# We are looking for conditional probability where the question is what the probability of B (Chlyamydia positive) given A (HPV positive) has already happened

# P(B given A) = P(B AND A) / P(A total)
# P(chl given hpv) = P(chl AND hpv)/P(hpv total)

# conditional probability
p.chl.given.hpv = p.hpv.chl/p.hpv.tot
p.chl.given.hpv
```


#### c. What is the probability that a woman who tests positive for Chlamydia will also test positive HPV?

```{r}
# set operations
# P(someone testing positive for HPV given that they have already tested positive for chlamydia)
# P(A given B) = P(B AND A)/P(B total)
# P(hpv given chl) = P(chl AND hpv)/P(hpv total)

# conditional probability
p.hpv.given.chl = (p.hpv.chl)/p.chl.tot
p.hpv.given.chl
```


#### d. What is the total probability that a woman from this population who gets tested would test positive for either HPV and/or Chlamydia?

```{r}
# P(A OR B) = (P(A) + P(B)) - P(A AND B)
p.chl.tot + p.hpv.tot - p.hpv.chl
```


#### e. Is the occurrence of these infections independent? Explain.

```{r}
# No, the occurrence of these infections are not independent because the chance of being positive for HPV if the patient already is infected with chlyamdia is very high at 60% which suggests that there is coninfection occuring. .
```


# 3. Allele frequencies

Many gene loci have a major allele and a number of minor alleles in the population. Let's say there are 5 alleles for a particular locus, that individuals mate randomly with respect to the genotype at this locus, and that allele frequencies for A1-A5 are 0.82, 0.06, 0.05, 0.04, and 0.03 respectively.

For each of the questions below, write out the probability equations (e.g. P(Ai) for each term) and then compute the answers numerically.

#### a. What is the probability that any single allele chosen at random from the population is either A2 or A5? Write out the probability statement as a comment below and then computer the answer.

```{r}
# Allele frequencies
A1 = 0.82
A2 = 0.06
A3 = 0.05
A4 = 0.04
A5 = 0.03

# Probability of either A2 or A5 = P(A2 OR A5) = P(A2) + P(A5)
A2 + A5
```


#### b. What is the probability that an individual carries two A3 alleles?

```{r}
# Probability that one individual carries two A3 alleles. A3 AND A3
A3 * A3
```


#### c. What is the probability that someone does NOT carry two A3 alleles?

```{r}
# Probability that someone does not carry two A3 alleles. That is the above answer subtracted from 1. 

1 - (A3 * A3)
```


#### d. What is the probability that someone is heterozygous for A1 and A4?

```{r}
# Probability that someone carries one A1 allele and 1 A4 allele. A1 AND A4
# We need to multiply by two because there are two possibilities of getting a heterozygote in a population (p^2 + 2pq + q^2 = 1)
2 * (A1 * A4)

```


#### e. What is the probability that *neither* of two random individuals in the population would carry two A1 alleles?

*Hint: first consider (i) the probability of carrying two A1 alleles, then (ii) the probability of NOT having two A1 alleles, and finally (iii) the probability that 2 individuals are not carrying two A1 alleles.*

```{r}
# prob of carrying two A1 alleles
A1 * A1

# prob of NOT having two A1 alleles
1 - (A1 * A1)

# P(2 individuals are NOT A1,A1)
(1 - (A1 * A1)) * (1 - A1 * A1)

```


#### f. *Extra credit:* What is the probability that 3 random individuals carry no A3 or A4 alleles at all? (Remember that each individual carries two alleles.) 

```{r}
# prob of an allele being A3 or A4
A3 + A4

# prob of an allele not being A3 or A4
one_allele <- 1 - (A3 + A4)
one_allele

# P(neither of 2 alleles is A3 or A4)
two_alleles <- one_allele * one_allele
two_alleles

# P(3 people have no A3 or A4 alleles)
three_alleles <- two_alleles^3
three_alleles

```


# 4. Trait inheritance

Some people are hypersensitive to the smell of asparagus, and can even detect a strong odor in the urine of a person who has recently eaten asparagus. This trait turns out to have a simple genetic basis (Online Mendelian Inheritance in Man, 2019):

+ An individual with one or two copies of the $A$ allele at the gene ($AA$ or $Aa$ genotypes) can smell asparagus in urine, whereas 
+ a person with two copies of the alternative $a$ allele ($aa$ genotypes) cannot. 

**In the human population, 5% of alleles are $A$ and 95% are $a$.**

Assume that men and women in the population have the same allele frequencies at the asparagus-smelling gene and that marriage and child production are independent of the genotype at the gene.


#### a. What is the probability that a randomly sampled individual from the population has two copies of the $a$ allele (that is, that it has an $aa$ genotype)? 

*Note: below we suggest to use a list to hold all of the different variables you will need to calculate, to help you keep track of the combinations. A list uses similar syntax to a data frame, except it can hold any kind of data of any size.*

```{r}
Pr=list()
Pr$a=0.95
Pr$A=0.05

Pr$aa = Pr$a * Pr$a
Pr$aa

Pr$AA = Pr$A * Pr$A
Pr$AA
```


#### b. What is the probability that both members of a randomly sampled married couple (man and woman) are $aa$ at the asparagus-smelling gene? 

```{r}
Pr$aa_aa = Pr$aa * Pr$aa
Pr$aa_aa
```


#### c. What is the probability that both members of a randomly sampled married couple (man and woman) are heterozygotes at this locus (meaning that each person has one allele $A$ and one allele $a$)? Remember that each allele can be inherited from either parent.

```{r}
# first calculate probability that one individual is heterozygous
# Multiplying by 2 because within a punnett square, there is a 50% chance, or two out of four possibilities to get a heterozygote.
Pr$Aa = 2*(Pr$A * Pr$a)
Pr$Aa

# probability that both parents are heterozygous
Pr$Aa_Aa = Pr$Aa * Pr$Aa
Pr$Aa_Aa        
```


#### d. What is the conditional probability that the first child of the couple in (c) is an $aa$ homozygote? Remember that the child must receive exactly one allele from each parent.

***Beware: here you need to consider transmission probabilities, not the allele frequencies in the population! Think "Punnett square".***

```{r collapse=TRUE}
# conditional probability that Aa x Aa -> aa
# probability that the first child is an aa homozygote given that the parents are both heterozygotes. Each parent must pass on their recessive allele and each parent has a 50% chance of doing that so the chances of both things happening is 0.25 because 0.5*0.5
Pr$Caa.FAa_MAa = 0.5 * 0.5
Pr$Caa.FAa_MAa
```


#### e. Above we figured out the probability that a child of known parentage inherits a particular allele combination. What if we want to turn this around and ask, ***If a randomly sampled child has the $aa$ genotype, what is the probability that both its parents were both $Aa$ heterozygotes?***

To answer this we need to use **Bayes' Theorem**. Recall the definition of conditional probability:

$$ P(F_{Aa} M_{Aa} | C_{aa}):= \frac{P(F_{Aa} M_{Aa} \cap C_{aa}) }{ P(C_{aa}) }$$
and

$$ P(C_{aa} | F_{Aa} M_{Aa}):= \frac{P(F_{Aa} M_{Aa} \cap C_{aa}) }{ P(F_{Aa} M_{Aa}) }$$


Therefore, the intersection (co-occurrence) of the parental and child alleles can be written equivalently as:

$$P(F_{Aa} M_{Aa} \cap C_{aa}) = P(F_{Aa} M_{Aa} | C_{aa}) * P(C_{aa}) = P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})$$

Fantastic! Now we have what we need to find the desired ***posterior probability*** $P(F_{Aa} M_{Aa} | C_{aa})$, given the known

+ ***priors*** $P(F_{alleles}M_{alleles})$ and
+ ***likelihoods*** $P(C_{aa}|F_{alleles}M_{alleles})$.

All we have to do is rearrange this equation and use Bayes Theorem:

$$P(F_{Aa} M_{Aa} | C_{aa}) = \frac{P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})}{P(C_{aa})} = \frac{P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})}{\sum_{ij} P(C_{aa}|F_iM_j) * P(F_iM_j)}$$

where $i,j \in \{AA, Aa, aA, aa\}$. 

This equation represents ***the fraction of the time that we expect to see one particular combination of parental alleles given an $aa$ child (i.e. both parents are $Aa$ heterozygotes), relative to all the possible ways a child could inherit two $a$ alleles.***

To compute the ***posterior probability***, we need to first quantify a bunch of individual probabilities, so let's break the problem into pieces and compute the required components in stages:

- **i. Probabilities of each allele combination in any individual in the population (parental frequencies):**

```{r collapse = TRUE}
# probabilities of each allele combination in the parents
Pr$AA =  Pr$A * Pr$A
Pr$Aa =  Pr$A * Pr$a
Pr$aA =  Pr$a * Pr$A
Pr$aa =  Pr$a * Pr$a

Pr$AA
Pr$Aa
Pr$aA
Pr$aa
```

- **ii. *Priors* for each combination of parental alleles (states of nature):**

```{r collapse=TRUE}
# priors for each combination of parental alleles
Pr$Faa_Maa = Pr$aa * Pr$aa
Pr$FAa_Maa = (Pr$Aa * Pr$aa) + (Pr$aA * Pr$aa)
Pr$FAA_Maa = Pr$AA * Pr$aa
Pr$Faa_MAa = (Pr$aa * Pr$Aa) + (Pr$aa * Pr$aA)
Pr$FAa_MAa = (Pr$Aa * Pr$Aa) + (Pr$aA * Pr$aA) + (Pr$Aa * Pr$aA) + (Pr$aA * Pr$Aa)
Pr$FAA_MAa = (Pr$AA * Pr$Aa) + (Pr$AA * Pr$aA)
Pr$Faa_MAA = Pr$aa * Pr$AA
Pr$FAa_MAA = (Pr$Aa * Pr$AA) + (Pr$aA * Pr$AA)
Pr$FAA_MAA = Pr$AA * Pr$AA

Pr$Faa_Maa
Pr$FAa_Maa
Pr$FAA_Maa
Pr$Faa_MAa
Pr$FAa_MAa
Pr$FAA_MAa
Pr$Faa_MAA
Pr$FAa_MAA
Pr$FAA_MAA
```

- **iii. *Likelihoods* (conditional probabilities) for $C_{aa}$ given each combination of parental alleles:**

```{r}
# individual likelihoods (think Punnett square)
Pr$Caa.Faa_Maa = 1
Pr$Caa.FAa_Maa = 0.5 * 1
Pr$Caa.FAA_Maa = 0 * 1
Pr$Caa.Faa_MAa = 1 * 0.5
Pr$Caa.FAa_MAa = 0.5 * 0.5
Pr$Caa.FAA_MAa = 0 * 0.5
Pr$Caa.Faa_MAA = 1 * 0
Pr$Caa.FAa_MAA = 0.5 * 0
Pr$Caa.FAA_MAA = 0 * 0
```


- **iv. *Total probability* $Pr(C_{aa})$ that the child is $aa$ (the denominator):**

To find this, unfortunately need to consider the probabilities of all possible combinations of parental allele contributions, or possible ***states of nature***:

$$\begin{aligned}P(C_{aa}) =& P(C_{aa}|F_{aa}M_{aa}) * P(F_{aa}M_{aa})  \ + \\
                            & P(C_{aa}|F_{Aa}M_{aa}) * P(F_{Aa}M_{aa})  \ + \\
                            & P(C_{aa}|F_{AA}M_{aa}) * P(F_{AA}M_{aa})  \ + \\
                            & P(C_{aa}|F_{aa}M_{Aa}) * P(F_{aa}M_{Aa})  \ + \\
                            & P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})  \ + \\
                            & P(C_{aa}|F_{AA}M_{Aa}) * P(F_{AA}M_{Aa})  \ + \\
                            & P(C_{aa}|F_{aa}M_{AA}) * P(F_{aa}M_{AA})  \ + \\
                            & P(C_{aa}|F_{Aa}M_{AA}) * P(F_{Aa}M_{AA})  \ + \\
                            & P(C_{aa}|F_{AA}M_{AA}) * P(F_{AA}M_{AA}) 
\end{aligned}$$


```{r}
# total probability P(C_Aa)
Pr$Caa = Pr$Caa.Faa_Maa * Pr$Faa_Maa +
    Pr$Caa.FAa_Maa * Pr$FAa_Maa +
    Pr$Caa.FAA_Maa * Pr$FAA_Maa +
    Pr$Caa.Faa_MAa * Pr$Faa_MAa +
    Pr$Caa.FAa_MAa * Pr$FAa_MAa +
    Pr$Caa.FAA_MAa * Pr$FAA_MAa +
    Pr$Caa.Faa_MAA * Pr$Faa_MAA +
    Pr$Caa.FAa_MAA * Pr$FAa_MAA +
    Pr$Caa.FAA_MAA * Pr$FAA_MAA
          

Pr$Caa
```


- **v. Finally, we can plug the desired variables from above into the original Bayes equation to find the *posterior probability:* **

$$P(F_{Aa}M_{Aa} | C_{aa}) = \frac{P(C_{aa}|F_{Aa}M_{Aa}) * P(F_{Aa}M_{Aa})}{P(C_{aa})}$$

```{r}
# Posterior probability of P(F_Aa, M_Aa | C_aa):
Pr$FAa_MAa.Caa = ((Pr$Caa.FAa_MAa) * (Pr$FAa_MAa))/Pr$Caa
Pr$FAa_MAa.Caa
```


#### f. If a randomly sampled child has the $aa$ genotype, what is the probability that both its parents were also $aa$?

The ***posterior probability*** $P(F_{aa} M_{aa} | C_{aa})$ is:

$$P(F_{aa} M_{aa} | C_{aa}) = \frac{P(C_{aa}|F_{aa}M_{aa}) * P(F_{aa}M_{aa})}{P(C_{aa})} = \frac{P(C_{aa}|F_{aa}M_{aa}) * P(F_{aa}M_{aa})}{\sum_{ij} P(C_{aa}|F_iM_j) * P(F_iM_j)}$$

We computed all the relevant quantities above, so now this should be easy!

```{r}
Pr$Faa_Maa.Caa = ((Pr$Caa | Pr$Faa_Maa) * (Pr$Faa_Maa))/Pr$Caa
Pr$Faa_Maa.Caa
```

---