---
title: "PCA/Clustering"
author: "Raza"
date: "12/1/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This dataset contains raw RNA-Seq counts from an experiment where cells were treated with 4 different drugs for either 1hr or 4hrs followed by RNA-Sequencing. Negative control cells (NC) were not treated with any drug. We will use clustering to analyse how the transcriptional landscape changes at different timepoints in response to different drugs. 

0. Load required libraries
```{r}
library(edgeR)
library(ComplexHeatmap)
library(ggplot2)
library(ggfortify)
```

1. Import csv file containing raw counts, remove rows with all zeros and convert to a matrix
```{r}
allcounts=read.csv("counts.csv",row.names = 1)
allcounts=allcounts[rowSums(allcounts)>0,]
allcounts=as.matrix(allcounts)

```

2. Apply cpm normalization (we are using cpm normalization just for practice, other normalizations such as rlog from deseq2 maybe preferable)
```{r}
allcounts=cpm(allcounts)

```

3. Make pca plot (use autoplot function from ggfortify)
```{r}
pr=prcomp(t(allcounts),scale=T,center = T)
autoplot(pr)

```

4. Color plot by sample group
```{r}
#make a dataframe containing sample names
metadata=data.frame(names=c("NC","NC","P_1","P_1","P_4","P_4","PF_4","PF_4","TP_4","TP_4","MT_4","MT_4"))

#use the dataframe with the autoplot function
autoplot(pr,data = metadata,col="names",size=5)

```

5. Make a plot showing percentage of variance explained by each PC (a scree plot)
```{r}

plot(pr)

#extract the s.d from the pca object and calculate variance 
v=(pr$sdev)^2
t=sum(v)
var_percent=v/t

#plot proportion of total variance against pc value
plot(var_percent,type = "b")

```

6. Shortlist variable genes for clustering
```{r}
#calculate avg expression for each gene and store in a variable
avg=apply(allcounts,1, mean)

#lets retain genes with avg expression of >10 cpm (this is an arbitrary threshold)
allcounts=allcounts[avg>10,] 

#calculate variance of each gene across all conditions and store in a variable
variance=apply(allcounts,1, var)

#sort genes by variance retain top 500 genes with the highest variance
allcounts=allcounts[order(variance,decreasing = T),]
top500=allcounts[1:500,]

```

7. Apply zscore scaling 
```{r}
#The scale function applies to columns by default. To scale genes we need to scale by row, so we transpose before and after scaling 
top500_scaled=t(scale(t(top500)))


```

8. Make a correlation matrix
```{r}
#The cor function applies to columns, so we transpose to get correlation between genes
c_mat=cor(t(top500_scaled))

```

9. Convert to a distance matrix
```{r}
#We subtract correlations from 1 so highly correlated genes are closest in terms of distance and vice versa
dist_mat=as.dist(1-c_mat)

```

10. Hierarchically cluster genes and make 4 clusters
```{r}
clust=hclust(dist_mat)
plot(clust)

#extract gene names from clusters if needed
test=data.frame(cutree(clust,k=4))


```

11. Cluster by kmeans and make 4 clusters 
```{r}
kclust=kmeans(dist_mat,4)

#extract gene names from clusters if needed
test_k=data.frame(kclust$cluster)

```


12. Make a heatmap of top500 variable genes clustered hierarchically using pearson correlation with the package "complexheatmaps"
```{r}
Heatmap(top500_scaled,show_row_names = F,name = "Heatmap",clustering_distance_rows = "pearson")

```

13. Make the same heatmap again without clustering columns
```{r}
Heatmap(top500_scaled,show_row_names = F,name = "Heatmap",clustering_distance_rows = "pearson",cluster_columns = F)


```

14. Make the same heatmap again but divide the genes into 4 cluster by kmeans
```{r}
Heatmap(top500_scaled,show_row_names = F,name = "Heatmap",clustering_distance_rows = "pearson",cluster_columns = F,row_km = 4)


```

15. Extract the genes lists from the different clusters
```{r}
hm=Heatmap(top500_scaled,show_row_names = F,name = "Heatmap",clustering_distance_rows = "pearson",cluster_columns = F,row_km = 4)

cluster1=rownames(top500_scaled)[row_order(hm)[[1]]]
cluster2=rownames(top500_scaled)[row_order(hm)[[2]]]
cluster3=rownames(top500_scaled)[row_order(hm)[[3]]]
cluster4=rownames(top500_scaled)[row_order(hm)[[4]]]
```


